{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "gpuClass": "standard",
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "50a6e508ca7940c197736c47944861c4": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DropdownModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DropdownModel",
            "_options_labels": [
              "None",
              "sp|P01308|INS_HUMAN",
              "sp|P01315|INS_PIG",
              "sp|P01317|INS_BOVIN",
              "sp|P01321|INS_CANLF",
              "sp|P01329|INS_CAVPO",
              "sp|P67970|INS_CHICK"
            ],
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "DropdownView",
            "description": "Select protein:",
            "description_tooltip": null,
            "disabled": false,
            "index": 4,
            "layout": "IPY_MODEL_4d616101f6eb4af0a4d4ff36c3e30595",
            "style": "IPY_MODEL_b0c06ab5be0643d88fc51414e9dc03e6"
          }
        },
        "4d616101f6eb4af0a4d4ff36c3e30595": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b0c06ab5be0643d88fc51414e9dc03e6": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/B2B_Tools_MSA_Example_Overall_behavior.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Bio2Byte - Multiple Sequence Alignment Analysis - Overall behavior.\n",
        "\n",
        "#### **Goal: Compute and visualize the biophysical and sequence conservation of a Multiple Sequence Alignment (MSA) using the b2btools.**\n",
        "\n",
        "*In this notebook we compute the biophysical and sequence conservation of the OVERALL BEHAVIOR of the MSA. To compute them in function of a reference protein, please refer to the following [notebook](https://colab.research.google.com/drive/1BBAGn4PCL9cqYaED5SRe4V8wXoeBibVq#scrollTo=guB0TE1iXZtD).*\n",
        "\n",
        "The following code will:\n",
        "0. Install the b2btools and its dependencies.\n",
        "1. Upload the Multiple Sequence Alignment (MSA) to study.\n",
        "2. Predict the biophysical properties of the input MSA:\n",
        "      *   Backbone and sidechain dynamics (Dynamine)\n",
        "      *   Conformational propensities (sheet, helix, coil, polyproline II) (Dynamine)\n",
        "      *   Early folding propensities (EFoldMine)\n",
        "      *   Disorder propensities (Disomine)\n",
        "3. Compute the sequence conservation (Shannon's entropy) of the input MSA at every postion as well as the MSA occupency.\n",
        "4. Compute 2D plots of the biophysical behaviour, the conservation of the biophysical properties and the sequence conservation.\n",
        "5. Compute the Gaussian Mixture Model scores of the proteins to identify residues with a different behavior compared to the other residues at the same position in the MSA.\n",
        "5. Download your results.\n",
        "\n",
        "**Check out our webserver: [online b2BTools](https://bio2byte.be/b2btools/)**"
      ],
      "metadata": {
        "id": "guB0TE1iXZtD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install b2bTools==3.0.7 py3dmol ipympl hmmer\n",
        "!pip install --upgrade matplotlib"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "Wg3_pDh4j-b9",
        "outputId": "c051da7e-9a29-40bc-81da-22cadec88ab9"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: b2bTools==3.0.7 in /usr/local/lib/python3.10/dist-packages (3.0.7)\n",
            "Requirement already satisfied: py3dmol in /usr/local/lib/python3.10/dist-packages (2.1.0)\n",
            "Requirement already satisfied: ipympl in /usr/local/lib/python3.10/dist-packages (0.9.4)\n",
            "Requirement already satisfied: hmmer in /usr/local/lib/python3.10/dist-packages (3.4.0.0)\n",
            "Requirement already satisfied: biopython~=1.83 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.83)\n",
            "Requirement already satisfied: Cython~=0.29.37 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (0.29.37)\n",
            "Collecting matplotlib~=3.5.3 (from b2bTools==3.0.7)\n",
            "  Using cached matplotlib-3.5.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.9 MB)\n",
            "Requirement already satisfied: numpy~=1.26.4 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.26.4)\n",
            "Requirement already satisfied: pandas~=1.5.3 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.5.3)\n",
            "Requirement already satisfied: pomegranate~=0.14 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (0.14.9)\n",
            "Requirement already satisfied: requests~=2.31.0 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (2.31.0)\n",
            "Requirement already satisfied: scikit-learn~=1.0.2 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.0.2)\n",
            "Requirement already satisfied: scipy~=1.12.0 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.12.0)\n",
            "Requirement already satisfied: torch~=1.13.1 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.13.1)\n",
            "Requirement already satisfied: torchvision~=0.14.1 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (0.14.1)\n",
            "Requirement already satisfied: urllib3~=1.26.6 in /usr/local/lib/python3.10/dist-packages (from b2bTools==3.0.7) (1.26.18)\n",
            "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.10/dist-packages (from ipympl) (0.2.0)\n",
            "Requirement already satisfied: ipython<9 in /usr/local/lib/python3.10/dist-packages (from ipympl) (7.34.0)\n",
            "Requirement already satisfied: ipywidgets<9,>=7.6.0 in /usr/local/lib/python3.10/dist-packages (from ipympl) (7.7.1)\n",
            "Requirement already satisfied: pillow in /usr/local/lib/python3.10/dist-packages (from ipympl) (9.4.0)\n",
            "Requirement already satisfied: traitlets<6 in /usr/local/lib/python3.10/dist-packages (from ipympl) (5.7.1)\n",
            "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (67.7.2)\n",
            "Requirement already satisfied: jedi>=0.16 in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (0.19.1)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (4.4.2)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (3.0.43)\n",
            "Requirement already satisfied: pygments in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (2.16.1)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (0.2.0)\n",
            "Requirement already satisfied: matplotlib-inline in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (0.1.7)\n",
            "Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.10/dist-packages (from ipython<9->ipympl) (4.9.0)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.10/dist-packages (from ipywidgets<9,>=7.6.0->ipympl) (5.5.6)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.10/dist-packages (from ipywidgets<9,>=7.6.0->ipympl) (3.6.6)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from ipywidgets<9,>=7.6.0->ipympl) (3.0.10)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.5.3->b2bTools==3.0.7) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.5.3->b2bTools==3.0.7) (4.51.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.5.3->b2bTools==3.0.7) (1.4.5)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.5.3->b2bTools==3.0.7) (24.0)\n",
            "Requirement already satisfied: pyparsing>=2.2.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.5.3->b2bTools==3.0.7) (3.1.2)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.5.3->b2bTools==3.0.7) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas~=1.5.3->b2bTools==3.0.7) (2023.4)\n",
            "Requirement already satisfied: joblib>=0.9.0b4 in /usr/local/lib/python3.10/dist-packages (from pomegranate~=0.14->b2bTools==3.0.7) (1.4.2)\n",
            "Requirement already satisfied: networkx>=2.4 in /usr/local/lib/python3.10/dist-packages (from pomegranate~=0.14->b2bTools==3.0.7) (3.3)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.10/dist-packages (from pomegranate~=0.14->b2bTools==3.0.7) (6.0.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests~=2.31.0->b2bTools==3.0.7) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests~=2.31.0->b2bTools==3.0.7) (3.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests~=2.31.0->b2bTools==3.0.7) (2024.2.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn~=1.0.2->b2bTools==3.0.7) (3.5.0)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.10/dist-packages (from torch~=1.13.1->b2bTools==3.0.7) (4.11.0)\n",
            "Requirement already satisfied: nvidia-cuda-runtime-cu11==11.7.99 in /usr/local/lib/python3.10/dist-packages (from torch~=1.13.1->b2bTools==3.0.7) (11.7.99)\n",
            "Requirement already satisfied: nvidia-cudnn-cu11==8.5.0.96 in /usr/local/lib/python3.10/dist-packages (from torch~=1.13.1->b2bTools==3.0.7) (8.5.0.96)\n",
            "Requirement already satisfied: nvidia-cublas-cu11==11.10.3.66 in /usr/local/lib/python3.10/dist-packages (from torch~=1.13.1->b2bTools==3.0.7) (11.10.3.66)\n",
            "Requirement already satisfied: nvidia-cuda-nvrtc-cu11==11.7.99 in /usr/local/lib/python3.10/dist-packages (from torch~=1.13.1->b2bTools==3.0.7) (11.7.99)\n",
            "Requirement already satisfied: wheel in /usr/local/lib/python3.10/dist-packages (from nvidia-cublas-cu11==11.10.3.66->torch~=1.13.1->b2bTools==3.0.7) (0.43.0)\n",
            "Requirement already satisfied: jupyter-client in /usr/local/lib/python3.10/dist-packages (from ipykernel>=4.5.1->ipywidgets<9,>=7.6.0->ipympl) (6.1.12)\n",
            "Requirement already satisfied: tornado>=4.2 in /usr/local/lib/python3.10/dist-packages (from ipykernel>=4.5.1->ipywidgets<9,>=7.6.0->ipympl) (6.3.3)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /usr/local/lib/python3.10/dist-packages (from jedi>=0.16->ipython<9->ipympl) (0.8.4)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.10/dist-packages (from pexpect>4.3->ipython<9->ipympl) (0.7.0)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.10/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython<9->ipympl) (0.2.13)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib~=3.5.3->b2bTools==3.0.7) (1.16.0)\n",
            "Requirement already satisfied: notebook>=4.4.1 in /usr/local/lib/python3.10/dist-packages (from widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (6.5.5)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (3.1.4)\n",
            "Requirement already satisfied: pyzmq<25,>=17 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (24.0.1)\n",
            "Requirement already satisfied: argon2-cffi in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (23.1.0)\n",
            "Requirement already satisfied: jupyter-core>=4.6.1 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (5.7.2)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (5.10.4)\n",
            "Requirement already satisfied: nbconvert>=5 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (6.5.4)\n",
            "Requirement already satisfied: nest-asyncio>=1.5 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.6.0)\n",
            "Requirement already satisfied: Send2Trash>=1.8.0 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.18.1)\n",
            "Requirement already satisfied: prometheus-client in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.20.0)\n",
            "Requirement already satisfied: nbclassic>=0.4.7 in /usr/local/lib/python3.10/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.0.0)\n",
            "Requirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.10/dist-packages (from jupyter-core>=4.6.1->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (4.2.2)\n",
            "Requirement already satisfied: jupyter-server>=1.8 in /usr/local/lib/python3.10/dist-packages (from nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.24.0)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.10/dist-packages (from nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.2.4)\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (4.9.4)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (4.12.3)\n",
            "Requirement already satisfied: bleach in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (6.1.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.7.1)\n",
            "Requirement already satisfied: entrypoints>=0.2.2 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.4)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.3.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (2.1.5)\n",
            "Requirement already satisfied: mistune<2,>=0.8.1 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.8.4)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.10.0)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.5.1)\n",
            "Requirement already satisfied: tinycss2 in /usr/local/lib/python3.10/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.3.0)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.10/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (2.19.1)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.10/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (4.19.2)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.10/dist-packages (from argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (21.2.0)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (23.2.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (2023.12.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.35.1)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.18.1)\n",
            "Requirement already satisfied: anyio<4,>=3.1.0 in /usr/local/lib/python3.10/dist-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (3.7.1)\n",
            "Requirement already satisfied: websocket-client in /usr/local/lib/python3.10/dist-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.8.0)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.16.0)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.10/dist-packages (from beautifulsoup4->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (2.5)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.10/dist-packages (from bleach->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (0.5.1)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.10/dist-packages (from anyio<4,>=3.1.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.3.1)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anyio<4,>=3.1.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (1.2.1)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.10/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets<9,>=7.6.0->ipympl) (2.22)\n",
            "Installing collected packages: matplotlib\n",
            "  Attempting uninstall: matplotlib\n",
            "    Found existing installation: matplotlib 3.0.0\n",
            "    Uninstalling matplotlib-3.0.0:\n",
            "      Successfully uninstalled matplotlib-3.0.0\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "bigframes 1.6.0 requires matplotlib>=3.7.1, but you have matplotlib 3.5.3 which is incompatible.\n",
            "bigframes 1.6.0 requires scikit-learn>=1.2.2, but you have scikit-learn 1.0.2 which is incompatible.\n",
            "plotnine 0.12.4 requires matplotlib>=3.6.0, but you have matplotlib 3.5.3 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed matplotlib-3.5.3\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "mpl_toolkits"
                ]
              },
              "id": "0ace3104210b4559bd583ea60afd3fc7"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (3.5.3)\n",
            "Collecting matplotlib\n",
            "  Downloading matplotlib-3.9.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.3 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.3/8.3 MB\u001b[0m \u001b[31m43.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.2.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (4.51.0)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.4.5)\n",
            "Requirement already satisfied: numpy>=1.23 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.26.4)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (24.0)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (9.4.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (3.1.2)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (2.8.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n",
            "Installing collected packages: matplotlib\n",
            "  Attempting uninstall: matplotlib\n",
            "    Found existing installation: matplotlib 3.5.3\n",
            "    Uninstalling matplotlib-3.5.3:\n",
            "      Successfully uninstalled matplotlib-3.5.3\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "b2btools 3.0.7 requires matplotlib~=3.5.3; python_version == \"3.10\", but you have matplotlib 3.9.0 which is incompatible.\n",
            "bigframes 1.6.0 requires scikit-learn>=1.2.2, but you have scikit-learn 1.0.2 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed matplotlib-3.9.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "wzSJQsPWYV8Z"
      },
      "outputs": [],
      "source": [
        "#@title 1. Install the Bio2Byte tools package and its dependencies\n",
        "\n",
        "#@markdown Please be patient, this can take several minutes.\n",
        "\n",
        "import math\n",
        "import Bio\n",
        "from Bio import AlignIO\n",
        "from sklearn import preprocessing\n",
        "import os\n",
        "import json\n",
        "import shutil\n",
        "from google.colab import files\n",
        "from b2bTools.multipleSeq.Predictor import MineSuiteMSA\n",
        "from google.colab import output\n",
        "import numpy as np\n",
        "from sklearn import mixture\n",
        "import sys\n",
        "from os import path"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 1. Upload your MSA file\n",
        "\n",
        "#@markdown Execute this cell to upload your MSA file from your computer.\n",
        "\n",
        "#@markdown **🚨⚠️ IMPORTANT input file size limits:**\n",
        "#@markdown * Dynamine (dynamics and secondary structure): min 5 residues per sequence\n",
        "#@markdown * Disomine (disorder): min 5 residues per sequence\n",
        "#@markdown * EfoldMine (early folding): min 5 residues per sequence\n",
        "\n",
        "#@markdown **🚨⚠️ IMPORTANT input file format:**\n",
        "#@markdown The file can be in CLUSTAL, FASTA, BaliBase, PSI, A3M, Blast, PHYLIP and STOCKHOLM format\n",
        "\n",
        "msa_filename = None\n",
        "\n",
        "uploaded = files.upload()\n",
        "\n",
        "for fn in uploaded.keys():\n",
        "  print('MSA file \"{name}\" with length {length} bytes uploaded with success'.format(\n",
        "      name=fn, length=len(uploaded[fn])))\n",
        "  msa_filename = fn\n",
        "\n",
        "\n",
        "#@markdown Select the format of your MSA file:\n",
        "fasta = True #@param {type:\"boolean\"}\n",
        "clustal = False #@param {type:\"boolean\"}\n",
        "phylip = False #@param {type:\"boolean\"}\n",
        "stockholm = False #@param {type:\"boolean\"}\n",
        "\n",
        "# BaliBase = False #@param {type:\"boolean\"}\n",
        "# PSI = False #@param {type:\"boolean\"}\n",
        "# A3M = False #@param {type:\"boolean\"}\n",
        "# Blast = False #@param {type:\"boolean\"}\n",
        "\n",
        "extensions=[fasta,clustal,phylip,stockholm]\n",
        "extensions_names=[\"fasta\",\"clustal\",\"phylip\",\"stockholm\"]\n",
        "for ext in extensions:\n",
        "  if ext:\n",
        "    extension = extensions_names[extensions.index(ext)]"
      ],
      "metadata": {
        "id": "7p6nFboEZHr2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 92
        },
        "outputId": "b7665129-5a8a-4b62-ece6-1a7ac689bead",
        "cellView": "form"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-49090c2e-f698-46e8-866d-5684591126f1\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-49090c2e-f698-46e8-866d-5684591126f1\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving INS_PROT.afa to INS_PROT (1).afa\n",
            "MSA file \"INS_PROT (1).afa\" with length 1107 bytes uploaded with success\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2. Select the predictors you want to use and run them\n",
        "#@markdown Select the bio2Byte tools you want to include into the predictions:\n",
        "%%capture\n",
        "\n",
        "#@markdown ### DynaMine predictor tool\n",
        "#@markdown >Fast predictor of protein backbone dynamics using only sequence information as input.\n",
        "#@markdown >The version here also predicts side-chain dynamics and secondary structure predictors\n",
        "#@markdown >using the same principle.\n",
        "\n",
        "#@markdown **Prediction values included**: `backbone`, `sidechain`, `helix`, `ppII`, `coil`, and `sheet`\n",
        "\n",
        "DynaMine = True #@param {type:\"boolean\"}\n",
        "\n",
        "#@markdown ### DisoMine predictor tool\n",
        "#@markdown >Predicts protein disorder with recurrent neural networks not directly\n",
        "#@markdown >from the amino acid sequence, but instead from more generic predictions of key\n",
        "#@markdown >biophysical properties, here protein dynamics, secondary structure\n",
        "#@markdown >and early folding.\n",
        "\n",
        "#@markdown **Prediction values included**: `disoMine`\n",
        "\n",
        "DisoMine = True #@param {type:\"boolean\"}\n",
        "\n",
        "#@markdown ### EFoldMine predictor tool\n",
        "#@markdown >Predicts from the primary amino acid sequence of a protein,\n",
        "#@markdown >which amino acids are likely involved in early folding events.\n",
        "\n",
        "#@markdown **Prediction values included**: `earlyFolding`\n",
        "\n",
        "EFoldMine = True #@param {type:\"boolean\"}\n",
        "\n",
        "#@markdown **Don't forget to run the cell after ticking the boxes.**\n",
        "\n",
        "msaSuite = MineSuiteMSA()\n",
        "\n",
        "if DynaMine and not DisoMine and not EFoldMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('dynamine'))\n",
        "elif DisoMine and not DynaMine and not EFoldMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('disoMine'))\n",
        "elif EFoldMine and not DynaMine and not DisoMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('eFoldMine'))\n",
        "\n",
        "elif DynaMine and DisoMine and not EFoldMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('disoMine', 'dynamine'))\n",
        "elif DynaMine and EFoldMine and not DisoMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('eFoldMine','dynamine'))\n",
        "elif DisoMine and EFoldMine and not DynaMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('eFoldMine', 'disoMine'))\n",
        "\n",
        "elif DisoMine and EFoldMine and DynaMine:\n",
        "  msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('eFoldMine', 'disoMine', 'dynamine'))\n",
        "\n",
        "msaSuite.getDistributions()\n",
        "jsondata_list = [msaSuite.alignedPredictionDistribs]"
      ],
      "metadata": {
        "id": "IsxI7K7srXkX",
        "cellView": "form"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.b Select a protein from the MSA to study its behavior on top of the behavior of the MSA\n",
        "\n",
        "#@markdown Select the protein you want to include in the study in the dropdown list that appears once you execute this cell.\n",
        "\n",
        "#@markdown In case you don't want to study a protein in particular, select None.\n",
        "\n",
        "\n",
        "import Bio\n",
        "from Bio import AlignIO\n",
        "import ipywidgets as widgets\n",
        "\n",
        "alignment_file=AlignIO.read( msa_filename, extension)\n",
        "\n",
        "ids =[]\n",
        "for prot in alignment_file:\n",
        "  ids.append(prot.id)\n",
        "\n",
        "out = widgets.Dropdown(\n",
        "    options=[\"None\"]+ids,\n",
        "    description='Select protein:',\n",
        "    disabled=False,\n",
        ")\n",
        "display(out)"
      ],
      "metadata": {
        "id": "aIEb5Gbokuch",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 49,
          "referenced_widgets": [
            "50a6e508ca7940c197736c47944861c4",
            "4d616101f6eb4af0a4d4ff36c3e30595",
            "b0c06ab5be0643d88fc51414e9dc03e6"
          ]
        },
        "cellView": "form",
        "outputId": "98568696-dca8-4bd6-de51-e467175de467"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Dropdown(description='Select protein:', options=('None', 'sp|P01308|INS_HUMAN', 'sp|P01315|INS_PIG', 'sp|P0131…"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "50a6e508ca7940c197736c47944861c4"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.c. Set the variables in function of the value selected above\n",
        "\n",
        "#selected protein in drop down list\n",
        "selected_prot = []\n",
        "\n",
        "try:\n",
        "  if out.value != \"None\":\n",
        "\n",
        "    selected_prot.append(out.value)\n",
        "\n",
        "    jsondata_list_selected = []\n",
        "    print(f\"The selected protein is {out.value}\")\n",
        "\n",
        "    #prediction selected protein\n",
        "    for prot in selected_prot:\n",
        "      predictions_single_seq = msaSuite.allAlignedPredictions\n",
        "      jsondata_list_selected.append([predictions_single_seq[prot]])\n",
        "except:\n",
        "  pass"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VyA8hle-LZZK",
        "outputId": "4ccbe17d-c660-403a-8cb4-68b3e5008ae5",
        "cellView": "form"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The selected protein is sp|P01321|INS_CANLF\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4.a. Sequence conservation (Shannon's entropy)\n",
        "#@markdown Once this cell has been executed, the conservation $C$ of the residues\n",
        "#@markdown at a particular column $x$ within the MSA is computed by using\n",
        "#@markdown the Shannon's entropy $E$ and by taking into account the number of gaps $G$ in that column:\n",
        "#@markdown $$ C(x) = (1- E(x))(1-G(x)) $$\n",
        "\n",
        "#@markdown Shannon's entropy at column $x$ equals:\n",
        "\n",
        "#@markdown $$ E(x) = - \\lambda \\sum_{a}^{K} p_{a} log_{2}(p_{a}) $$\n",
        "\n",
        "#@markdown where $K$, the alphabet size which equals 21 as it includes the 20 amino acid types and 1 symbol for the gaps,\n",
        "\n",
        "#@markdown and $ p_{a}$ is the probability of observing the *ath* symbol type at position $x$.\n",
        "\n",
        "#@markdown $\\lambda$ scales the entropy to range [0,1]:\n",
        "\n",
        "#@markdown if the number of residue types in column $x$, $M$ ($1\\leq M \\leq K$) equals 1:\n",
        "#@markdown $$ \\lambda = 1 $$\n",
        "#@markdown else:\n",
        "#@markdown $$ \\lambda = [log_{2}(M)]^{-1} $$\n",
        "\n",
        "#@markdown Moreover, the number of gaps in column $x$:\n",
        "#@markdown $$ G(x) = \\dfrac{n}{N} $$\n",
        "#@markdown where $n$ is the number of gaps in column $x$.\n",
        "\n",
        "#Shannon_entropy calculates the conservation of the amino acids at each position. It takes into account\n",
        "#the gaps in the columns. Check https://doi.org/10.1002/prot.10146 for info formula\n",
        "def shannon_entropy(list_input):\n",
        "    tot = len(list_input) #total number of AA at particular position in MSA\n",
        "    gaps = list_input.count(\"-\") / tot #count frequency of gaps in that column of MSA\n",
        "    unique_base = set(list_input) #remove duplicates, \"-\" is seen as an amino acid type\n",
        "    unique_base_len = len(unique_base) #total number of AA at particular position in MSA with no duplicates\n",
        "\n",
        "\n",
        "    entropy_list = [] # entropy of AA at particular position\n",
        "\n",
        "    for base in unique_base:\n",
        "        n_i = list_input.count(base)\n",
        "        P_i = n_i / tot\n",
        "\n",
        "        entropy_i = P_i * (math.log(P_i, 2))\n",
        "        entropy_list.append(entropy_i)\n",
        "\n",
        "    #sum entropy of every residue at a position and normalize it so it is between 0 and 1\n",
        "    entropy_sum = math.fsum(entropy_list)\n",
        "\n",
        "    if unique_base_len == 1: # log(1, 2) = 0; n/0 throws ZeroDivisionError\n",
        "        shannon_entropy = entropy_sum\n",
        "    else:\n",
        "        unique_base_len_log = math.log(unique_base_len, 2)\n",
        "        shannon_entropy = (-1 / unique_base_len_log) * entropy_sum\n",
        "\n",
        "    #Return entropy AA and entropy gaps at 1 position\n",
        "    #If entropy is high than there are many possible arrangements (high variability)\n",
        "    return shannon_entropy, gaps\n",
        "\n",
        "def conservation(alignment_file):\n",
        "    conservation_AA_list = []\n",
        "\n",
        "    for col_no in range(len(list(alignment_file[0]))):\n",
        "        list_input = list(alignment_file[:, col_no])\n",
        "\n",
        "        sh_entropy_AA, sh_entropy_gaps = shannon_entropy(list_input)\n",
        "\n",
        "        #Translate entropy into conservation. 0 means no conservation, 1 highly conserved\n",
        "        #As e took into account the gaps: if only gaps: entropy=0 freq_gaps=1 conservation = 0\n",
        "        conservation_AA = (1 - sh_entropy_AA) * (1 - sh_entropy_gaps)\n",
        "        conservation_AA_list.append(conservation_AA)\n",
        "\n",
        "    return conservation_AA_list\n",
        "\n",
        "#Read MSA\n",
        "alignment_file = AlignIO.read(msa_filename, extension)\n",
        "conservation_AA_list = conservation(alignment_file)"
      ],
      "metadata": {
        "id": "DDZ7JQfqaMhi"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4.b. Sequence occupency in MSA\n",
        "#@markdown Once this cell has been executed, the sequence occupency in the MSA is computed (i.e., the frequency of gaps at every residue position in the MSA)\n",
        "\n",
        "counter = [0] * alignment_file.get_alignment_length()\n",
        "total_seq = len(alignment_file)\n",
        "for prot in alignment_file:\n",
        "  selected_prot_seq=list(prot.seq)\n",
        "  for position,residue in enumerate(selected_prot_seq):\n",
        "      if residue != \"-\":\n",
        "          counter[position] +=1\n",
        "freq_gap = [i/total_seq for i in counter]"
      ],
      "metadata": {
        "id": "WheiI8YN6sBl"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5. Prepare to plot results\n",
        "#@markdown Run this cell in order to prepare the notebook context to render different plots. Please be patient this can take a few minutes.\n",
        "%matplotlib widget\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "import os\n",
        "import math\n",
        "\n",
        "if not os.path.exists(\"/content/results\"):\n",
        "    os.mkdir(\"/content/results\", )\n",
        "\n",
        "if DynaMine and not DisoMine and not EFoldMine:\n",
        "    NB_SUBPLOTS = 6\n",
        "    PREDICTION_TITLES = {\n",
        "      'backbone': \"DynaMine backbone dynamics\",\n",
        "      'sidechain': \"DynaMine sidechain dynamics\",\n",
        "      'ppII': \"DynaMine conformational propensities: ppII (polyproline II)\",\n",
        "      'coil': \"DynaMine conformational propensities: Coil\",\n",
        "      'sheet': \"DynaMine conformational propensities: Sheet\",\n",
        "      'helix': \"DynaMine conformational propensities: Helix\",\n",
        "  }\n",
        "\n",
        "    PREDICTION_POSITION = {\n",
        "        'backbone':     0,\n",
        "        'sidechain':    1,\n",
        "        'ppII':         2,\n",
        "        'coil':         3,\n",
        "        'sheet':        4,\n",
        "        'helix':        5,\n",
        "    }\n",
        "\n",
        "elif DisoMine and not DynaMine and not EFoldMine:\n",
        "  NB_SUBPLOTS = 1\n",
        "  PREDICTION_TITLES = {\n",
        "       'disoMine': \"Disorder (disoMine)\"\n",
        "  }\n",
        "\n",
        "  PREDICTION_POSITION = {\n",
        "      'disoMine': 0\n",
        "  }\n",
        "\n",
        "elif EFoldMine and not DisoMine and not DynaMine:\n",
        "  NB_SUBPLOTS = 1\n",
        "\n",
        "  PREDICTION_TITLES = {\n",
        "      'earlyFolding': \"Early folding (EFoldMine)\"\n",
        "  }\n",
        "\n",
        "  PREDICTION_POSITION = {\n",
        "      'earlyFolding': 0\n",
        "  }\n",
        "\n",
        "elif DynaMine and DisoMine and not EFoldMine:\n",
        "  NB_SUBPLOTS = 7\n",
        "  PREDICTION_TITLES = {\n",
        "      'backbone': \"DynaMine backbone dynamics\",\n",
        "      'sidechain': \"DynaMine sidechain dynamics\",\n",
        "      'ppII': \"DynaMine conformational propensities: ppII (polyproline II)\",\n",
        "      'coil': \"DynaMine conformational propensities: Coil\",\n",
        "      'sheet': \"DynaMine conformational propensities: Sheet\",\n",
        "      'helix': \"DynaMine conformational propensities: Helix\",\n",
        "      'disoMine': \"Disorder (disoMine)\"\n",
        "  }\n",
        "\n",
        "  PREDICTION_POSITION = {\n",
        "      'backbone':     0,\n",
        "      'sidechain':    1,\n",
        "      'ppII':         2,\n",
        "      'coil':         3,\n",
        "      'sheet':        4,\n",
        "      'helix':        5,\n",
        "      'disoMine':     6\n",
        "  }\n",
        "\n",
        "elif DynaMine and EFoldMine and not DisoMine:\n",
        "  NB_SUBPLOTS = 7\n",
        "  PREDICTION_TITLES = {\n",
        "      'backbone': \"DynaMine backbone dynamics\",\n",
        "      'sidechain': \"DynaMine sidechain dynamics\",\n",
        "      'ppII': \"DynaMine conformational propensities: ppII (polyproline II)\",\n",
        "      'coil': \"DynaMine conformational propensities: Coil\",\n",
        "      'sheet': \"DynaMine conformational propensities: Sheet\",\n",
        "      'helix': \"DynaMine conformational propensities: Helix\",\n",
        "      'earlyFolding': \"Early folding (EFoldMine)\"\n",
        "  }\n",
        "\n",
        "  PREDICTION_POSITION = {\n",
        "      'backbone':     0,\n",
        "      'sidechain':    1,\n",
        "      'ppII':         2,\n",
        "      'coil':         3,\n",
        "      'sheet':        4,\n",
        "      'helix':        5,\n",
        "      'earlyFolding': 6\n",
        "  }\n",
        "elif EFoldMine and DisoMine and not DynaMine:\n",
        "  NB_SUBPLOTS = 2\n",
        "  PREDICTION_TITLES = {\n",
        "      'earlyFolding': \"Early folding (EFoldMine)\",\n",
        "      'disoMine': \"Disorder (disoMine)\"\n",
        "  }\n",
        "\n",
        "  PREDICTION_POSITION = {\n",
        "      'earlyFolding': 0,\n",
        "      'disoMine':     1\n",
        "  }\n",
        "elif DynaMine and DisoMine and EFoldMine:\n",
        "  NB_SUBPLOTS = 8\n",
        "  PREDICTION_TITLES = {\n",
        "      'backbone': \"DynaMine backbone dynamics\",\n",
        "      'sidechain': \"DynaMine sidechain dynamics\",\n",
        "      'ppII': \"DynaMine conformational propensities: ppII (polyproline II)\",\n",
        "      'coil': \"DynaMine conformational propensities: Coil\",\n",
        "      'sheet': \"DynaMine conformational propensities: Sheet\",\n",
        "      'helix': \"DynaMine conformational propensities: Helix\",\n",
        "      'earlyFolding': \"Early folding (EFoldMine)\",\n",
        "      'disoMine': \"Disorder (disoMine)\"\n",
        "  }\n",
        "\n",
        "  PREDICTION_POSITION = {\n",
        "      'backbone':     0,\n",
        "      'sidechain':    1,\n",
        "      'ppII':         2,\n",
        "      'coil':         3,\n",
        "      'sheet':        4,\n",
        "      'helix':        5,\n",
        "      'earlyFolding': 6,\n",
        "      'disoMine':     7\n",
        "  }\n",
        "\n",
        "AXIS_TITLES = {\n",
        "    \"x\": \"Residue position in the MSA\",\n",
        "    \"y\": \"Prediction values\"\n",
        "}\n",
        "\n",
        "\n",
        "import matplotlib as mpl\n",
        "mpl.use(\"Agg\")\n",
        "import matplotlib.pyplot as plt\n",
        "from matplotlib.ticker import AutoMinorLocator\n",
        "import matplotlib.ticker as ticker\n"
      ],
      "metadata": {
        "id": "85bOAJZ7n142"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.a. Plot the predicted biophysical conservation of the MSA\n",
        "#@markdown Once this cell has been executed, a file with plots will be saved in the folder \"Results\"\n",
        "#@markdown In case you want to zoom in on the plots, run the cell 5.b below .\n",
        "%%capture\n",
        "\n",
        "#@markdown The plots represent the biophysical behaviour of the proteins in the MSA.\n",
        "#@markdown There is one graph for every biophyisical property that is studied.\n",
        "#@markdown As the median, 1st and 3rd quartile and outliers are shown on the graphs,\n",
        "#@markdown the conservation of the biophysical properties can be observed.\n",
        "\n",
        "#@markdown The dots on the plots show the sequence occupency of the MSA, there are 4 different sizes representing 25%,50%,75% and 100% of the sequence occupency.\n",
        "\n",
        "# from matplotlib.collections import LineCollection\n",
        "\n",
        "def plot_biophysical_msa(jsondata_list_interest,jsondata_list_selected,selected_prots,sequences,freq_gap):\n",
        "\n",
        "    colors = ['blue', 'orange']\n",
        "    residues_count = len(jsondata_list_interest[0]['backbone']['median'])\n",
        "    sequences_count = len(sequences)\n",
        "    #Plot representation\n",
        "    fig, axs = plt.subplots(NB_SUBPLOTS)\n",
        "    fig.set_figwidth(20)\n",
        "    fig.set_figheight(50)\n",
        "\n",
        "    plt.suptitle(f\"Predicted biophysical properties of the MSA: {residues_count} aligned residues from {sequences_count} sequences\", fontsize=14)\n",
        "\n",
        "    # These for loops got too complicated, I have to think\n",
        "    # something simpler to handle the None values in the data\n",
        "    predictions = jsondata_list_interest[0].keys()\n",
        "    for prediction_index, biophys_data in enumerate(predictions):\n",
        "        if biophys_data == 'agmata':\n",
        "            continue\n",
        "\n",
        "        subplot_index_row = PREDICTION_POSITION[biophys_data]\n",
        "\n",
        "        ax = axs[subplot_index_row]\n",
        "        for data, col in zip(jsondata_list_interest, colors):\n",
        "            none_idx = []\n",
        "\n",
        "            for n in range(residues_count):\n",
        "                if data[biophys_data]['median'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'firstQuartile'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'thirdQuartile'][n] == None:\n",
        "                    none_idx.append(n)\n",
        "\n",
        "            range_list = []\n",
        "            for n in range(len(none_idx)):\n",
        "                try:\n",
        "                    if none_idx[n] + 1 != none_idx[n + 1]:\n",
        "                        range_list.append(\n",
        "                            (none_idx[n] + 1, none_idx[n + 1]))\n",
        "                    else:\n",
        "                        continue\n",
        "                except:\n",
        "                    if len(none_idx) == 1:\n",
        "                        range_list.append((0, none_idx[0]))\n",
        "                        range_list.append((none_idx[0] + 1, len(\n",
        "                            data[biophys_data][\n",
        "                                'median'])))\n",
        "\n",
        "                    else:\n",
        "                        range_list.append((0, none_idx[0]))\n",
        "                        range_list.append((none_idx[-1] + 1, len(\n",
        "                            data[biophys_data][\n",
        "                                'median'])))\n",
        "\n",
        "            # When there are None values in the data\n",
        "            if range_list:\n",
        "                for tuple in range_list:\n",
        "                    x = np.arange(tuple[0], tuple[1], 1)\n",
        "                    firstq = \\\n",
        "                        data[biophys_data][\n",
        "                            'firstQuartile'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    thirdq = \\\n",
        "                        data[biophys_data][\n",
        "                            'thirdQuartile'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    bottom = \\\n",
        "                        data[biophys_data][\n",
        "                            'bottomOutlier'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    top = \\\n",
        "                        data[biophys_data]['topOutlier'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    ax.fill_between(\n",
        "                        x, firstq, thirdq, alpha=0.3, color=col, label=\"1st-3rd Quartiles\")\n",
        "                    ax.fill_between(\n",
        "                        x, bottom, top, alpha=0.1, color=col, label=\"Outliers\")\n",
        "\n",
        "            # When there aren't None values in the data\n",
        "            else:\n",
        "                x = np.arange(0, len(\n",
        "                    data[biophys_data]['median']), 1)\n",
        "                firstq = data[biophys_data][\n",
        "                    'firstQuartile']\n",
        "                thirdq = data[biophys_data][\n",
        "                    'thirdQuartile']\n",
        "                bottom = data[biophys_data][\n",
        "                    'bottomOutlier']\n",
        "                top = data[biophys_data]['topOutlier']\n",
        "                ax.fill_between(\n",
        "                    x, firstq, thirdq, alpha=0.3, color=col, label=\"1st-3rd Quartiles\")\n",
        "                ax.fill_between(\n",
        "                    x, bottom, top, alpha=0.1, color=col, label=\"Outliers\")\n",
        "\n",
        "            ax.plot(data[biophys_data]['median'], linewidth=1.25, color=col, label=\"Median\")\n",
        "\n",
        "            #Add the selected protein if there are some\n",
        "            colors_2 = [\"magenta\", \"blue\", \"cyan\"] #adapt in function of number of proteins of interest, now max 3 can be studied simultanously\n",
        "            if len(selected_prot)>0:\n",
        "              for count,prot in enumerate(selected_prot):\n",
        "                  ax.plot(jsondata_list_selected[count][0][biophys_data], '-x', linewidth=1.5, color=colors_2[count], label=f\"Prediction {prot}\")\n",
        "\n",
        "            #To represent the amount of sequences at every column of the MSA\n",
        "            limits = [0.25,0.5,0.75,1]\n",
        "            limit_previous = 0\n",
        "            size = 20\n",
        "            for limit in limits:\n",
        "                plot_x = []\n",
        "                plot_y = []\n",
        "                for idx, freq in enumerate(freq_gap):\n",
        "                    if limit_previous < freq <= limit:\n",
        "                        plot_x.append(idx)\n",
        "                        plot_y.append(data[biophys_data]['median'][idx])\n",
        "                ax.scatter(plot_x,plot_y,marker='o',s=size, color=col)\n",
        "                limit_previous = limit\n",
        "                size += 15\n",
        "\n",
        "        ax.set_title(PREDICTION_TITLES[biophys_data])\n",
        "        ax.axis([0, residues_count-1, min(bottom)-0.05, max(top)+0.05])\n",
        "        # #labels\n",
        "        # labels = []\n",
        "        # count = 1\n",
        "        # while count < residues_count:\n",
        "        #     labels.append(str(count))\n",
        "        #     count +=4\n",
        "        # ax.xaxis.set_ticks(ticks=np.arange(0,residues_count+1,4))\n",
        "        # ax.xaxis.set_ticklabels(labels)\n",
        "        # ax.xaxis.set_minor_locator(AutoMinorLocator())\n",
        "        # ax.yaxis.set_minor_locator(AutoMinorLocator())\n",
        "        ax.set_ylabel(AXIS_TITLES['y'])\n",
        "        ax.set_xlabel(AXIS_TITLES['x'])\n",
        "        if biophys_data == 'backbone':\n",
        "            ax.axhline(y=1.0, color='green', linewidth= 1.5, linestyle='-.', label=\"Above: Membrane spaning\") #Membrane spaning\n",
        "            ax.axhline(y=0.8, color='orange', linewidth= 1.5, linestyle='-.', label=\"Above: Rigid\") #Membrane spaning\n",
        "            if min(bottom)-0.05 < 0.69:\n",
        "                ax.axhline(y=0.69, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Context dependent \\nBelow: Flexible\") #context dependent (either rigide or flexible)\n",
        "        if biophys_data == 'earlyFolding':\n",
        "            ax.axhline(y=0.169, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Likely to start folding\") #above: likely start protein folding process\n",
        "        if biophys_data == 'disoMine':\n",
        "            ax.axhline(y=0.5, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Likely to be disordered\") #above: likely disordered\n",
        "        ax.legend(ncol=1, bbox_to_anchor =(1.01,0.5), loc='center left')\n",
        "\n",
        "    plt.tight_layout()\n",
        "    fig.subplots_adjust(top=0.96, hspace = 0.2)\n",
        "    plt.savefig('/content/results/msa_biophysical_conservation.pdf')\n",
        "    plt.show()\n",
        "\n",
        "    return fig, axs\n",
        "\n",
        "sequences =  msaSuite.seqs\n",
        "fig, axs = plot_biophysical_msa(jsondata_list,jsondata_list_selected,selected_prot,sequences,freq_gap)"
      ],
      "metadata": {
        "id": "rmjYOzRzYfA6"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.b. Zoom in: Plot the predicted biophysical conservation of the MSA\n",
        "#@markdown Same as cell 5.a. but with zooms on specific regions of the MSA.\n",
        "%%capture\n",
        "\n",
        "#@markdown Enter in how many sections you want to cut your plots:\n",
        "\n",
        "n = 5 #@param {type:\"integer\"}\n",
        "#@markdown\n",
        "\n",
        "#@markdown Once this cell has been executed, a file with plots will be saved in the folder \"Results\".\n",
        "\n",
        "#@markdown The plots represent the biophysical behaviour of the proteins in the MSA.\n",
        "#@markdown There is one graph for every biophyisical property that is studied.\n",
        "#@markdown As the median, 1st and 3rd quartile and outliers are shown on the graphs,\n",
        "#@markdown the conservation of the biophysical properties can be observed.\n",
        "\n",
        "#@markdown The dots on the plots show the sequence occupency of the MSA, there are 4 different sizes representing 25%,50%,75% and 100% of the sequence occupency.\n",
        "\n",
        "#create sections\n",
        "residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "all_regions = []\n",
        "for elem in range(1,n+1):\n",
        "    label = \"Section%d_of_%d\" %(elem, n)\n",
        "    low = int(residues_count * ((1/n)*(elem-1)))\n",
        "    high = int(residues_count * ((1/n)*elem))\n",
        "    if all_regions == []:\n",
        "      low = 1\n",
        "    all_regions.append((label, (low, high)))\n",
        "\n",
        "def plot_biophysical_msa_zoom(jsondata_list_interest,jsondata_list_selected,selected_prot, sequences,freq_gap,all_regions):\n",
        "    colors = ['blue', 'orange']\n",
        "    residues_count = len(jsondata_list_interest[0]['backbone']['median'])\n",
        "    sequences_count = len(sequences)\n",
        "\n",
        "    for region in all_regions:\n",
        "        name_region = region[0]\n",
        "        lower_lim = region[1][0]\n",
        "        upper_lim =region[1][1]\n",
        "\n",
        "        #Plot representation\n",
        "        fig, axs = plt.subplots(NB_SUBPLOTS)\n",
        "        fig.set_figwidth(20)\n",
        "        fig.set_figheight(50)\n",
        "\n",
        "        fig.suptitle(f'Predicted biophysical properties of the MSA: \"{name_region}\" (residue {lower_lim} to residue {upper_lim-1}) from {sequences_count} sequences', fontsize=14)\n",
        "\n",
        "        # These for loops got too complicated, I have to think\n",
        "        # something simpler to handle the None values in the data\n",
        "        predictions = jsondata_list_interest[0].keys()\n",
        "        for prediction_index, biophys_data in enumerate(predictions):\n",
        "            if biophys_data == 'agmata':\n",
        "                continue\n",
        "\n",
        "            subplot_index_row = PREDICTION_POSITION[biophys_data]\n",
        "\n",
        "            ax = axs[subplot_index_row]\n",
        "            for data, col in zip(jsondata_list_interest, colors):\n",
        "                none_idx = []\n",
        "\n",
        "                for n in range(residues_count):\n",
        "                    if data[biophys_data]['median'][n] == None \\\n",
        "                            or data[biophys_data][\n",
        "                        'firstQuartile'][n] == None \\\n",
        "                            or data[biophys_data][\n",
        "                        'thirdQuartile'][n] == None:\n",
        "                        none_idx.append(n)\n",
        "\n",
        "                range_list = []\n",
        "                for n in range(len(none_idx)):\n",
        "                    try:\n",
        "                        if none_idx[n] + 1 != none_idx[n + 1]:\n",
        "                            range_list.append(\n",
        "                                (none_idx[n] + 1, none_idx[n + 1]))\n",
        "                        else:\n",
        "                            continue\n",
        "                    except:\n",
        "                        if len(none_idx) == 1:\n",
        "                            range_list.append((0, none_idx[0]))\n",
        "                            range_list.append((none_idx[0] + 1, len(\n",
        "                                data[biophys_data][\n",
        "                                    'median'])))\n",
        "\n",
        "                        else:\n",
        "                            range_list.append((0, none_idx[0]))\n",
        "                            range_list.append((none_idx[-1] + 1, len(\n",
        "                                data[biophys_data][\n",
        "                                    'median'])))\n",
        "\n",
        "                # When there are None values in the data\n",
        "                if range_list:\n",
        "                    for tuple in range_list:\n",
        "                        x = np.arange(tuple[0], tuple[1], 1)\n",
        "                        firstq = \\\n",
        "                            data[biophys_data][\n",
        "                                'firstQuartile'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        thirdq = \\\n",
        "                            data[biophys_data][\n",
        "                                'thirdQuartile'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        bottom = \\\n",
        "                            data[biophys_data][\n",
        "                                'bottomOutlier'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        top = \\\n",
        "                            data[biophys_data]['topOutlier'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        ax.fill_between(\n",
        "                            x[lower_lim:upper_lim], firstq[lower_lim:upper_lim], thirdq[lower_lim:upper_lim], alpha=0.3, color=col, label=\"1st & 3rd quartile\")\n",
        "                        ax.fill_between(\n",
        "                            x[lower_lim:upper_lim], bottom[lower_lim:upper_lim], top[lower_lim:upper_lim], alpha=0.1, color=col, label=\"Outliers\")\n",
        "\n",
        "                # When there aren't None values in the data\n",
        "                else:\n",
        "                    x = np.arange(lower_lim,upper_lim, 1)\n",
        "                    firstq = data[biophys_data][\n",
        "                        'firstQuartile'][lower_lim:upper_lim]\n",
        "                    thirdq = data[biophys_data][\n",
        "                        'thirdQuartile'][lower_lim:upper_lim]\n",
        "                    bottom = data[biophys_data][\n",
        "                        'bottomOutlier'][lower_lim:upper_lim]\n",
        "                    top = data[biophys_data]['topOutlier'][lower_lim:upper_lim]\n",
        "                    ax.fill_between(\n",
        "                        x, firstq, thirdq, alpha=0.3, color=col, label=\"1st & 3rd quartile\")\n",
        "                    ax.fill_between(\n",
        "                        x, bottom, top, alpha=0.1, color=col, label=\"Outliers\")\n",
        "\n",
        "                    ax.plot(data[biophys_data]['median'],linewidth=1.25, color=\"black\", label= \"Median\")\n",
        "\n",
        "                    #Add selected proteins if there are some\n",
        "                    colors_2 = [\"magenta\", \"blue\", \"cyan\"] #adapt in function of number of proteins of interest, now max 3 can be studied simultanously\n",
        "                    if len(selected_prot)>0:\n",
        "                      for count,prot in enumerate(selected_prot):\n",
        "                          ax.plot(jsondata_list_selected[count][0][biophys_data], '-x', linewidth=1.5, color=colors_2[count], label=f\"Prediction {prot}\")\n",
        "\n",
        "                #To represent the amount of sequences occupency at every column of the MSA\n",
        "                limits = [0.25,0.5,0.75,1] #we divide the occupency in 4 groups: i<25%, 25%<i<50%, 50%<i<75%, i>75%\n",
        "                #depending on the group, the size of the dot will change\n",
        "                limit_previous = 0\n",
        "                size = 20\n",
        "                for limit in limits:\n",
        "                    plot_x = []\n",
        "                    plot_y = []\n",
        "                    for idx, freq in enumerate(freq_gap):\n",
        "                        if limit_previous < freq <= limit:\n",
        "                            plot_x.append(idx)\n",
        "                            plot_y.append(data[biophys_data]['median'][idx])\n",
        "                    if limit == 1:\n",
        "                      ax.scatter(plot_x,plot_y,marker='o',s=size, color=\"grey\", label = \"Size dot ~ freq. seq. occupency\")\n",
        "                    else:\n",
        "                      ax.scatter(plot_x,plot_y,marker='o',s=size, color=\"grey\")\n",
        "                    limit_previous = limit\n",
        "                    size += 15\n",
        "\n",
        "            ax.set_title(PREDICTION_TITLES[biophys_data])\n",
        "            ax.axis([lower_lim,upper_lim, min(bottom)-0.05, max(top)+0.05])\n",
        "            ax.set_ylabel(AXIS_TITLES['y'])\n",
        "            ax.set_xlabel(AXIS_TITLES['x'])\n",
        "            # ax.set_xticks(list(np.arange(lower_lim,upper_lim,10)))\n",
        "            if biophys_data == 'backbone':\n",
        "                ax.axhline(y=1.0, color='green', linewidth= 1.25, linestyle='-.', label=\"Above: Membrane spaning\") #Membrane spaning\n",
        "                ax.axhline(y=0.8, color='orange', linewidth= 1.25, linestyle='-.', label=\"Above: Rigid\") #Membrane spaning\n",
        "                if min(bottom)-0.05 < 0.69:\n",
        "                    ax.axhline(y=0.69, color='red', linewidth= 1.25, linestyle='-.', label=\"Above: Context dependent \\nBelow: Flexible\") #context dependent (either rigide or flexible)\n",
        "            if biophys_data == 'earlyFolding':\n",
        "                ax.axhline(y=0.169, color='red', linewidth= 1.25, linestyle='-.', label=\"Above: Likely to start folding\") #above: likely start protein folding process\n",
        "            if biophys_data == 'disoMine':\n",
        "                ax.axhline(y=0.5, color='red', linewidth= 1.25, linestyle='-.', label=\"Above: Likely to be disordered\") #above: likely disordered\n",
        "            ax.legend(ncol=1, bbox_to_anchor =(1.01,0.5), loc='center left')\n",
        "\n",
        "\n",
        "        plt.tight_layout()\n",
        "        fig.subplots_adjust(top=0.96, hspace = 0.2)\n",
        "        plt.savefig('/content/results/'+ name_region + '_msa_biophysical_conservation.pdf')\n",
        "        plt.show()\n",
        "\n",
        "    return fig, axs\n",
        "\n",
        "sequences =  msaSuite.seqs\n",
        "fig, axs = plot_biophysical_msa_zoom(jsondata_list,jsondata_list_selected,selected_prot, sequences,freq_gap,all_regions)"
      ],
      "metadata": {
        "id": "c-DONQkRghzm"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.c. Plot the biophysical and sequence conservation of the MSA\n",
        "#@markdown Once this cell has been executed, a file with plots will be saved in the folder \"Results\"\n",
        "#@markdown In case you want to zoom in on the plots, run the cell 5.d below.\n",
        "%%capture\n",
        "\n",
        "#@markdown The plots represent the biophysical behaviour of the proteins in the MSA as well as their sequence conservation.\n",
        "#@markdown There is one graph for every biophyisical property that is studied.\n",
        "#@markdown As the median, 1st and 3rd quartile and outliers are shown on the graphs,\n",
        "#@markdown the conservation of the biophysical properties can be observed.\n",
        "\n",
        "#@markdown The dots on the plots show the sequence occupency of the MSA, there are 4 different sizes representing 25%,50%,75% and 100% of the sequence occupency.\n",
        "\n",
        "def plot_entropy_biphysical_msa(jsondata_list, jsondata_list_selected, selected_prot, sequences, conservation_AA_list, freq_gap):\n",
        "    sequences_count = len(sequences)\n",
        "    residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "\n",
        "    #Color map for conservation sequence\n",
        "    cmap = mpl.cm.Blues(np.linspace(0,1,20))\n",
        "    cmap = mpl.colors.ListedColormap(cmap[:5])\n",
        "    colors = ['blue', 'orange']\n",
        "\n",
        "    #Plot representation\n",
        "    fig, axs = plt.subplots(NB_SUBPLOTS)\n",
        "    fig.set_figwidth(20)\n",
        "    fig.set_figheight(50)\n",
        "\n",
        "    fig.suptitle(f\"Predicted biophysical properties of the MSA: {residues_count} aligned residues from {len(sequences)} sequences\", fontsize=16)\n",
        "\n",
        "    predictions = jsondata_list[0].keys()\n",
        "    for prediction_index, biophys_data in enumerate(predictions):\n",
        "        if biophys_data == 'agmata':\n",
        "            continue\n",
        "\n",
        "        subplot_index_row = PREDICTION_POSITION[biophys_data]\n",
        "\n",
        "        ax = axs[subplot_index_row]\n",
        "        for data, col in zip(jsondata_list, colors):\n",
        "            none_idx = []\n",
        "\n",
        "            for n in range(residues_count):\n",
        "                if data[biophys_data]['median'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'firstQuartile'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'thirdQuartile'][n] == None:\n",
        "                    none_idx.append(n)\n",
        "\n",
        "            range_list = []\n",
        "            for n in range(len(none_idx)):\n",
        "                try:\n",
        "                    if none_idx[n] + 1 != none_idx[n + 1]:\n",
        "                        range_list.append(\n",
        "                            (none_idx[n] + 1, none_idx[n + 1]))\n",
        "                    else:\n",
        "                        continue\n",
        "                except:\n",
        "                    if len(none_idx) == 1:\n",
        "                        range_list.append((0, none_idx[0]))\n",
        "                        range_list.append((none_idx[0] + 1, len(\n",
        "                            data[biophys_data][\n",
        "                                'median'])))\n",
        "\n",
        "                    else:\n",
        "                        range_list.append((0, none_idx[0]))\n",
        "                        range_list.append((none_idx[-1] + 1, len(\n",
        "                            data[biophys_data][\n",
        "                                'median'])))\n",
        "\n",
        "            # When there are None values in the data\n",
        "            if range_list:\n",
        "                for tuple in range_list:\n",
        "                    x = np.arange(tuple[0], tuple[1], 1)\n",
        "                    firstq = \\\n",
        "                        data[biophys_data][\n",
        "                            'firstQuartile'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    thirdq = \\\n",
        "                        data[biophys_data][\n",
        "                            'thirdQuartile'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    bottom = \\\n",
        "                        data[biophys_data][\n",
        "                            'bottomOutlier'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    top = \\\n",
        "                        data[biophys_data]['topOutlier'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    ax.plot(x,firstq, linewidth=0.5, color=\"black\", label=\"1st & 3rd quartile\")\n",
        "                    ax.plot(x,thirdq, linewidth=0.5, color=\"black\")\n",
        "                    ax.plot(x,bottom, alpha=0.25, linestyle =\"--\", color=\"black\", label=\"Bottom & Top outliers\")\n",
        "                    ax.plot(x,top, alpha=0.25, linestyle =\"--\", color=\"black\")\n",
        "\n",
        "            # When there aren't None values in the data\n",
        "            else:\n",
        "                x = np.arange(0, len(\n",
        "                    data[biophys_data]['median']), 1)\n",
        "                firstq = data[biophys_data][\n",
        "                    'firstQuartile']\n",
        "                thirdq = data[biophys_data][\n",
        "                    'thirdQuartile']\n",
        "                bottom = data[biophys_data][\n",
        "                    'bottomOutlier']\n",
        "                top = data[biophys_data]['topOutlier']\n",
        "\n",
        "                ax.plot(x,firstq, linewidth=0.5, color=\"black\", label=\"1st & 3rd quartile\")\n",
        "                ax.plot(x,thirdq, linewidth=0.5, color=\"black\")\n",
        "                ax.plot(x,bottom, alpha=0.25, linestyle =\"--\", color=\"black\", label=\"Bottom & Top outliers\")\n",
        "                ax.plot(x,top, alpha=0.25, linestyle =\"--\", color=\"black\")\n",
        "\n",
        "                ax.plot(data[biophys_data]['median'], linewidth=1.25, color=\"black\", label= \"Median\")\n",
        "\n",
        "                #Add selected proteins if there are some\n",
        "                colors_2 = [\"magenta\", \"blue\", \"cyan\"] #adapt in function of number of proteins of interest, now max 3 can be studied simultanously\n",
        "                if len(selected_prot)>0:\n",
        "                  for count,prot in enumerate(selected_prot):\n",
        "                      ax.plot(jsondata_list_selected[count][0][biophys_data], '-x', linewidth=1.5, color=colors_2[count], label=f\"Prediction {prot}\")\n",
        "\n",
        "                #plot sequence conservation\n",
        "                entropy_values = conservation_AA_list\n",
        "\n",
        "                extent = [0, residues_count, min(bottom)-0.05, max(top)+0.05]\n",
        "                ax.imshow(np.array(entropy_values)[np.newaxis,:], cmap=cmap, extent=extent,aspect = \"auto\", vmin=0, vmax=1)\n",
        "                cbar = ax.figure.colorbar(\n",
        "                    mpl.cm.ScalarMappable(norm=None, cmap=cmap), shrink=1, pad = 0.03, ax=ax, ticks=[0, 1])\n",
        "                cbar.ax.set_yticklabels(['0%', '100%'])\n",
        "                cbar.ax.set_ylabel('Sequence conservation', rotation=270)\n",
        "\n",
        "            #To represent the amount of sequences occupency at every column of the MSA\n",
        "            limits = [0.25,0.5,0.75,1] #we divide the occupency in 4 groups: i<25%, 25%<i<50%, 50%<i<75%, i>75%\n",
        "            #depending on the group, the size of the dot will change\n",
        "            limit_previous = 0\n",
        "            size = 20\n",
        "            for limit in limits:\n",
        "                plot_x = []\n",
        "                plot_y = []\n",
        "                for idx, freq in enumerate(freq_gap):\n",
        "                    if limit_previous < freq <= limit:\n",
        "                        plot_x.append(idx)\n",
        "                        plot_y.append(data[biophys_data]['median'][idx])\n",
        "                if limit == 1:\n",
        "                  ax.scatter(plot_x,plot_y,marker='o',s=size, color=\"grey\", label = \"Size dot ~ freq. seq. occupency\")\n",
        "                else:\n",
        "                  ax.scatter(plot_x,plot_y,marker='o',s=size, color=\"grey\")\n",
        "                limit_previous = limit\n",
        "                size += 15\n",
        "\n",
        "        ax.set_title(PREDICTION_TITLES[biophys_data] , fontsize=14)\n",
        "        ax.axis([0, residues_count, min(bottom)-0.05, max(top)+0.05])\n",
        "        # ax.xaxis.set_major_locator(ticker.FixedLocator(np.arange(0,residues_count,4)))\n",
        "\n",
        "        # #labels\n",
        "        # labels = []\n",
        "        # count = 1\n",
        "        # while count < residues_count+1:\n",
        "        #     labels.append(str(count))\n",
        "        #     count +=4\n",
        "        # ax.xaxis.set_major_formatter(ticker.FixedFormatter(labels))\n",
        "        # ax.xaxis.set_minor_locator(ticker.FixedLocator(np.arange(0.99,residues_count+0.99,1)))\n",
        "        # ax.yaxis.set_minor_locator(AutoMinorLocator())\n",
        "        ax.set_ylabel(AXIS_TITLES['y'], fontsize=10)\n",
        "        ax.set_xlabel(AXIS_TITLES['x'], fontsize=10)\n",
        "\n",
        "        if biophys_data == 'backbone':\n",
        "            ax.axhline(y=1.0, color='green', linewidth= 1.5, linestyle='-.', label=\"Above: Membrane spaning\") #Membrane spaning\n",
        "            ax.axhline(y=0.8, color='orange', linewidth= 1.5, linestyle='-.', label=\"Above: Rigid\") #Membrane spaning\n",
        "            if min(bottom)-0.05 < 0.69:\n",
        "                ax.axhline(y=0.69, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Context dependent \\nBelow: Flexible\") #context dependent (either rigide or flexible)\n",
        "        if biophys_data == 'earlyFolding':\n",
        "            ax.axhline(y=0.169, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Likely to start folding\") #above: likely start protein folding process\n",
        "        if biophys_data == 'disoMine':\n",
        "            ax.axhline(y=0.5, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Likely to be disordered\") #above: likely disordered\n",
        "        ax.legend(ncol=1, bbox_to_anchor =(1.1,0.5), loc='center left')\n",
        "\n",
        "\n",
        "    plt.tight_layout()\n",
        "    fig.subplots_adjust(top=0.96, hspace = 0.6)\n",
        "\n",
        "    plt.savefig('/content/results/msa_sequence_biophysical_conservation.pdf')\n",
        "\n",
        "    return fig, axs\n",
        "\n",
        "sequences =  msaSuite.seqs\n",
        "fig, axs = plot_entropy_biphysical_msa(jsondata_list,jsondata_list_selected,selected_prot,sequences,conservation_AA_list,freq_gap)"
      ],
      "metadata": {
        "id": "tBTnNwxoc--c"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.d. Zoom in: Plot the sequence and biophysical conservation of the MSA\n",
        "#@markdown Same as cell 5.c. but with zooms on specific regions of the MSA.\n",
        "%%capture\n",
        "\n",
        "#@markdown Enter in how many sections you want to cut your plots:\n",
        "n = 5 #@param {type:\"integer\"}\n",
        "\n",
        "#@markdown\n",
        "\n",
        "#@markdown The plots represent the biophysical behaviour of the proteins in the MSA as well as their sequence conservation.\n",
        "#@markdown There is one graph for every biophyisical property that is studied.\n",
        "#@markdown As the median, 1st and 3rd quartile and outliers are shown on the graphs,\n",
        "#@markdown the conservation of the biophysical properties can be observed.\n",
        "\n",
        "#@markdown The dots on the plots show the sequence occupency of the MSA, there are 4 different sizes representing 25%,50%,75% and 100% of the sequence occupency.\n",
        "\n",
        "#create sections\n",
        "residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "all_regions = []\n",
        "for elem in range(1,n+1):\n",
        "    label = \"Section%d_of_%d\" %(elem, n)\n",
        "    low = int(residues_count * ((1/n)*(elem-1)))\n",
        "    high = int(residues_count * ((1/n)*elem))\n",
        "    if all_regions == []:\n",
        "      low = 1\n",
        "    all_regions.append((label, (low, high)))\n",
        "print(all_regions)\n",
        "#@markdown The dots on the plots show the sequence occupency of the MSA, there are 4 different sizes representing 25%,50%,75% and 100% of the sequence occupency.\n",
        "\n",
        "def plot_entropy_biphysical_msa_zoom(jsondata_list,jsondata_list_selected,selected_prot, sequences, all_regions, conservation_AA_list, freq_gap):\n",
        "    sequences_count = len(sequences)\n",
        "    residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "\n",
        "    #Color map for conservation sequence\n",
        "    cmap = mpl.cm.Blues(np.linspace(0,1,20))\n",
        "    cmap = mpl.colors.ListedColormap(cmap[:5])\n",
        "\n",
        "    colors = ['blue', 'orange']\n",
        "\n",
        "    for region in all_regions:\n",
        "        name_region = region[0]\n",
        "        lower_lim = region[1][0]\n",
        "        upper_lim = region[1][1]\n",
        "\n",
        "        #Plot representation\n",
        "        fig, axs = plt.subplots(8)\n",
        "        fig.set_figwidth(20)\n",
        "        fig.set_figheight(50)\n",
        "\n",
        "        fig.suptitle(f'Predicted biophysical properties of the MSA: \"{name_region}\" (residue {lower_lim} to residue {upper_lim-1}) from {sequences_count} sequences', fontsize=14)\n",
        "\n",
        "        # These for loops got too complicated, I have to think\n",
        "        # something simpler to handle the None values in the data\n",
        "        predictions = jsondata_list[0].keys()\n",
        "        for prediction_index, biophys_data in enumerate(predictions):\n",
        "            if biophys_data == 'agmata':\n",
        "                continue\n",
        "\n",
        "            subplot_index_row = PREDICTION_POSITION[biophys_data]\n",
        "\n",
        "            ax = axs[subplot_index_row]\n",
        "            for data, col in zip(jsondata_list, colors):\n",
        "                none_idx = []\n",
        "\n",
        "                for n in range(residues_count):\n",
        "                    if data[biophys_data]['median'][n] == None \\\n",
        "                            or data[biophys_data][\n",
        "                        'firstQuartile'][n] == None \\\n",
        "                            or data[biophys_data][\n",
        "                        'thirdQuartile'][n] == None:\n",
        "                        none_idx.append(n)\n",
        "\n",
        "                range_list = []\n",
        "                for n in range(len(none_idx)):\n",
        "                    try:\n",
        "                        if none_idx[n] + 1 != none_idx[n + 1]:\n",
        "                            range_list.append(\n",
        "                                (none_idx[n] + 1, none_idx[n + 1]))\n",
        "                        else:\n",
        "                            continue\n",
        "                    except:\n",
        "                        if len(none_idx) == 1:\n",
        "                            range_list.append((0, none_idx[0]))\n",
        "                            range_list.append((none_idx[0] + 1, len(\n",
        "                                data[biophys_data][\n",
        "                                    'median'])))\n",
        "\n",
        "                        else:\n",
        "                            range_list.append((0, none_idx[0]))\n",
        "                            range_list.append((none_idx[-1] + 1, len(\n",
        "                                data[biophys_data][\n",
        "                                    'median'])))\n",
        "\n",
        "                # When there are None values in the data\n",
        "                if range_list:\n",
        "                    for tuple in range_list:\n",
        "                        x = np.arange(tuple[0], tuple[1], 1)\n",
        "                        firstq = \\\n",
        "                            data[biophys_data][\n",
        "                                'firstQuartile'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        thirdq = \\\n",
        "                            data[biophys_data][\n",
        "                                'thirdQuartile'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        bottom = \\\n",
        "                            data[biophys_data][\n",
        "                                'bottomOutlier'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        top = \\\n",
        "                            data[biophys_data]['topOutlier'][\n",
        "                            tuple[0]:tuple[1]]\n",
        "                        ax.plot(x,firstq, linewidth=0.25, color=\"black\", label=\"1st & 3rd quartile\")\n",
        "                        ax.plot(x,thirdq, linewidth=0.25, color=\"black\")\n",
        "                        ax.plot(x,bottom, alpha=0.25, linestyle =\"--\", color=\"black\", label=\"Bottom & Top outliers\")\n",
        "                        ax.plot(x,top, alpha=0.25, linestyle =\"--\", color=\"black\")\n",
        "\n",
        "                # When there aren't None values in the data\n",
        "                else:\n",
        "                    x = np.arange(0, len(\n",
        "                        data[biophys_data]['median']), 1)\n",
        "                    firstq = data[biophys_data][\n",
        "                        'firstQuartile']\n",
        "                    thirdq = data[biophys_data][\n",
        "                        'thirdQuartile']\n",
        "                    bottom = data[biophys_data][\n",
        "                        'bottomOutlier']\n",
        "                    top = data[biophys_data]['topOutlier']\n",
        "\n",
        "                    ax.plot(x,firstq, linewidth=0.5, color=\"black\", label=\"1st & 3rd quartile\")\n",
        "                    ax.plot(x,thirdq, linewidth=0.5, color=\"black\")\n",
        "                    ax.plot(x,bottom, alpha=0.25, linestyle =\"--\", color=\"black\", label=\"Bottom & Top outliers\")\n",
        "                    ax.plot(x,top, alpha=0.25, linestyle =\"--\", color=\"black\")\n",
        "\n",
        "                    ax.plot(data[biophys_data]['median'], linewidth=1.25, color=\"black\", label= \"Median\")\n",
        "\n",
        "                    #Add selected proteins if there are some\n",
        "                    colors_2 = [\"magenta\", \"blue\", \"cyan\"] #adapt in function of number of proteins of interest, now max 3 can be studied simultanously\n",
        "                    if len(selected_prot)>0:\n",
        "                      for count,prot in enumerate(selected_prot):\n",
        "                          ax.plot(jsondata_list_selected[count][0][biophys_data], '-x', linewidth=1.5, color=colors_2[count], label=f\"Prediction {prot}\")\n",
        "\n",
        "                    #Sequence conservation\n",
        "                    entropy_values = conservation_AA_list\n",
        "                    extent = [0, residues_count, min(bottom)-0.05, max(top)+0.05]\n",
        "                    ax.imshow(np.array(entropy_values)[np.newaxis,:], cmap=cmap, extent=extent,aspect = \"auto\", vmin=0, vmax=1)\n",
        "                    cbar = ax.figure.colorbar(\n",
        "                        mpl.cm.ScalarMappable(norm=None, cmap=cmap), shrink=1, pad = 0.03, ax=ax, ticks=[0, 1])\n",
        "                    cbar.ax.set_yticklabels(['0%', '100%'])\n",
        "                    cbar.ax.set_ylabel('Sequence conservation', rotation=270)\n",
        "\n",
        "                #To represent the amount of sequences at every column of the MSA\n",
        "                limits = [0.25,0.5,0.75,1]\n",
        "                limit_previous = 0\n",
        "                size = 20\n",
        "                for limit in limits:\n",
        "                    plot_x = []\n",
        "                    plot_y = []\n",
        "                    for idx, freq in enumerate(freq_gap):\n",
        "                        if limit_previous < freq <= limit:\n",
        "                            plot_x.append(idx)\n",
        "                            plot_y.append(data[biophys_data]['median'][idx])\n",
        "                    if limit == 1:\n",
        "                      ax.scatter(plot_x,plot_y,marker='o',s=size, color=\"grey\", label = \"Size dot ~ freq. seq. occupency\")\n",
        "                    else:\n",
        "                      ax.scatter(plot_x,plot_y,marker='o',s=size, color=\"grey\")\n",
        "                    limit_previous = limit\n",
        "                    size += 15\n",
        "\n",
        "            ax.set_title(PREDICTION_TITLES[biophys_data] , fontsize=14)\n",
        "            ax.axis([lower_lim, upper_lim, min(bottom)-0.05, max(top)+0.05])\n",
        "            # ax.xaxis.set_major_locator(ticker.FixedLocator(np.arange(lower_lim,upper_lim-1,4)))\n",
        "            # #labels\n",
        "            # labels = []\n",
        "            # count = lower_lim + 1\n",
        "            # while count < upper_lim+1:\n",
        "            #     labels.append(str(count))\n",
        "            #     count +=4\n",
        "            # ax.xaxis.set_major_formatter(ticker.FixedFormatter(labels))\n",
        "            # ax.xaxis.set_minor_locator(ticker.FixedLocator(np.arange(0.99,residues_count+0.99,1)))\n",
        "            # ax.yaxis.set_minor_locator(AutoMinorLocator())\n",
        "            ax.set_ylabel(AXIS_TITLES['y'], fontsize=10)\n",
        "            ax.set_xlabel(AXIS_TITLES['x'], fontsize=10)\n",
        "\n",
        "\n",
        "            if biophys_data == 'backbone':\n",
        "                ax.axhline(y=1.0, color='green', linewidth= 1.5, linestyle='-.', label=\"Above: Membrane spaning\") #Membrane spaning\n",
        "                ax.axhline(y=0.8, color='orange', linewidth= 1.5, linestyle='-.', label=\"Above: Rigid\") #Membrane spaning\n",
        "                if min(bottom)-0.05 < 0.69:\n",
        "                    ax.axhline(y=0.69, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Context dependent \\nBelow: Flexible\") #context dependent (either rigide or flexible)\n",
        "            if biophys_data == 'earlyFolding':\n",
        "                ax.axhline(y=0.169, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Likely to start folding\") #above: likely start protein folding process\n",
        "            if biophys_data == 'disoMine':\n",
        "                ax.axhline(y=0.5, color='red', linewidth= 1.5, linestyle='-.', label=\"Above: Likely to be disordered\") #above: likely disordered\n",
        "            ax.legend(ncol=1, bbox_to_anchor =(1.1,0.5), loc='center left')\n",
        "\n",
        "\n",
        "        plt.tight_layout()\n",
        "        fig.subplots_adjust(top=0.96, hspace = 0.6)\n",
        "\n",
        "        plt.savefig('/content/results/'+ name_region + '_msa_sequence_biophysical_conservation.pdf')\n",
        "\n",
        "        plt.show()\n",
        "\n",
        "    return fig, axs\n",
        "\n",
        "sequences =  msaSuite.seqs\n",
        "fig, axs = plot_entropy_biphysical_msa_zoom(jsondata_list,jsondata_list_selected,selected_prot, sequences, all_regions, conservation_AA_list, freq_gap)"
      ],
      "metadata": {
        "id": "8YTLpmnqgTgZ"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 6. Get the Gaussian Mixture Model (GMM) scores\n",
        "\n",
        "#@markdown Once this cell has been executed, the GMM scores will be computed and the most different residues will be shown.\n",
        "#@markdown The results are also saved in the Results folder.\n",
        "\n",
        "#@markdown Note: only the following biophysical properties are considered to calculate\n",
        "#@markdown the GMM scores: backbone & sidechain dynamics, coil, sheet & helix conformation propensity,\n",
        "#@markdown early folding propsensity and disorder propensity.\n",
        "\n",
        "def GMM_scores (jsondata_list, single_preds):\n",
        "    #Train predictor\n",
        "    data_full = []\n",
        "    for aln_pos in range(len(jsondata_list[0]['backbone']['median'])):\n",
        "        pred_vector = []\n",
        "        for biophys in ['backbone', 'sidechain', 'coil', 'sheet', 'helix',\n",
        "                        'earlyFolding', 'disoMine']:\n",
        "            pred_vector.append(jsondata_list[0][biophys]['median'][aln_pos])\n",
        "        data_full.append(pred_vector)\n",
        "    X_train = np.vstack(data_full)\n",
        "    clf = mixture.GaussianMixture(n_components = 1, covariance_type = 'full',\n",
        "                                      verbose = 2, verbose_interval = 1)\n",
        "    clf.fit(X_train)\n",
        "\n",
        "    #Predict GMM scores\n",
        "    gmm_dict = {}\n",
        "    scores = []\n",
        "    gmm_info = {}\n",
        "    sequences = {}\n",
        "\n",
        "    for prot in single_preds.keys():\n",
        "      if prot != \"sequence\":\n",
        "        full_pred = []\n",
        "        for res in range(len(single_preds[prot]['backbone'])):\n",
        "            pred_vector = []\n",
        "            for biophys in ['backbone', 'sidechain', 'coil', 'sheet', 'helix',\n",
        "                            'earlyFolding', \"disoMine\"]:\n",
        "                if single_preds[prot][biophys][res] != None:\n",
        "                    pred_vector.append(single_preds[prot][biophys][res])\n",
        "\n",
        "            full_pred.append(pred_vector)\n",
        "\n",
        "        full_pred = [list for list in full_pred if list]\n",
        "\n",
        "        preds = np.vstack(full_pred)\n",
        "        gmm_info[prot] = clf.score_samples(preds).tolist()\n",
        "        scores.extend(clf.score_samples(preds).tolist())\n",
        "        gmm_info[prot] = clf.score_samples(preds).tolist()\n",
        "        scores.extend(clf.score_samples(preds).tolist())\n",
        "\n",
        "\n",
        "    for prot in predictions_single_seq[\"sequence\"].keys():\n",
        "      sequences[prot]=predictions_single_seq[\"sequence\"][prot]\n",
        "\n",
        "    #most different biophysical behaviour\n",
        "    perc_5 = np.percentile(scores, 5)\n",
        "    perc_1 = np.percentile(scores, 1)\n",
        "\n",
        "    for key in gmm_info.keys():\n",
        "\n",
        "        worst_5perc = []\n",
        "        worst_1perc = []\n",
        "\n",
        "        #find the residues that are the most different compared to the other residues at the same position in the MSA\n",
        "        #we analyze the difference in biophysical behavior between the residues\n",
        "        for i in gmm_info[key]:\n",
        "            if i <= perc_5:\n",
        "                idx = gmm_info[key].index(i)\n",
        "                AA = sequences[key][idx]\n",
        "                worst_5perc.append(AA+str(idx+1))\n",
        "            if i <= perc_1:\n",
        "                idx = gmm_info[key].index(i)\n",
        "                AA = sequences[key][idx]\n",
        "                worst_1perc.append(AA+str(idx+1))\n",
        "\n",
        "        #To avoid to have the same AA in both lists\n",
        "        for AA in worst_5perc:\n",
        "            if AA in worst_1perc:\n",
        "                worst_5perc.remove(AA)\n",
        "\n",
        "        worst = [worst_5perc,worst_1perc]\n",
        "        new_list = []\n",
        "        for worst_ in worst:\n",
        "            consecutive = []\n",
        "            for position in range(len(worst_)-1):\n",
        "                idx = worst_[position][1:]\n",
        "                next = worst_[position+1][1:]\n",
        "                if int(idx) +1 == int(next):\n",
        "                    consecutive.append(True) #True means there is a consecutive one coming\n",
        "                else:\n",
        "                    consecutive.append(False)\n",
        "            consecutive.append(False)\n",
        "\n",
        "            found_first = False\n",
        "            new = []\n",
        "            for i in range(len(worst_)):\n",
        "                if consecutive[i] == True and found_first == False:\n",
        "                    found_first = True\n",
        "                    AA_first = worst_[i]\n",
        "                if consecutive[i] == False and found_first == True:\n",
        "                    AA_last = worst_[i]\n",
        "                    concatenated = AA_first+\"-\"+AA_last\n",
        "                    new.append(concatenated)\n",
        "                    found_first = False\n",
        "                elif consecutive[i] == False and found_first == False:\n",
        "                    new.append(worst_[i])\n",
        "            new_list.append(new)\n",
        "        worst_5perc_new = new_list[0]\n",
        "        worst_1perc_new = new_list[1]\n",
        "\n",
        "\n",
        "        #Worst 5% means the residues are lower than the 5th percentille\n",
        "        #Worst 1% means the residues are lower than the 1th percentill\n",
        "        gmm_dict[key] = {\"GMMscore\": gmm_info[key], \"Worst 5%\":worst_5perc_new, \"Worst 1%\": worst_1perc_new}\n",
        "\n",
        "    with open('/content/results/GMM_quantification.json', 'w') as fp:\n",
        "        json.dump(gmm_dict, fp)\n",
        "\n",
        "    return gmm_dict\n",
        "\n",
        "jsondata_list = [msaSuite.alignedPredictionDistribs]\n",
        "single_preds = msaSuite.allAlignedPredictions\n",
        "gmm_scores = GMM_scores(jsondata_list, single_preds)\n",
        "\n",
        "\n",
        "print(\"\\n\")\n",
        "print(\"GMM score analysis to find the residues that are the most different:\")\n",
        "for prot in gmm_scores.keys():\n",
        "    print(prot)\n",
        "    print(\"Worst 5%:\")\n",
        "    print(str(gmm_scores[prot][\"Worst 5%\"]))\n",
        "    print(\"Worst 1%:\")\n",
        "    print(str(gmm_scores[prot][\"Worst 1%\"]))\n",
        "    print(\"\\n\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "RSi9TK3DwSpT",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b454dd00-db70-4a86-afe6-8bc79cba32df"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Initialization 0\n",
            "  Iteration 1\t time lapse 0.06274s\t ll change inf\n",
            "  Iteration 2\t time lapse 0.00755s\t ll change 0.00000\n",
            "Initialization converged: True\t time lapse 0.07042s\t ll 8.24536\n",
            "\n",
            "\n",
            "GMM score analysis to find the residues that are the most different:\n",
            "sp|P01308|INS_HUMAN\n",
            "Worst 5%:\n",
            "['M1-A2', 'G70', 'P72-G73', 'Q105', 'Y109']\n",
            "Worst 1%:\n",
            "['G71-P72']\n",
            "\n",
            "\n",
            "sp|P01315|INS_PIG\n",
            "Worst 5%:\n",
            "['G71']\n",
            "Worst 1%:\n",
            "[]\n",
            "\n",
            "\n",
            "sp|P01317|INS_BOVIN\n",
            "Worst 5%:\n",
            "['L13-L14', 'Y40', 'G76-L77', '-79']\n",
            "Worst 1%:\n",
            "['G75-G76', '-78']\n",
            "\n",
            "\n",
            "sp|P01321|INS_CANLF\n",
            "Worst 5%:\n",
            "['G73-G75']\n",
            "Worst 1%:\n",
            "[]\n",
            "\n",
            "\n",
            "sp|P01329|INS_CAVPO\n",
            "Worst 5%:\n",
            "['M1', 'A15-W17', 'P19-T21', 'Q23', 'Y40']\n",
            "Worst 1%:\n",
            "['G18', 'S41']\n",
            "\n",
            "\n",
            "sp|P67970|INS_CHICK\n",
            "Worst 5%:\n",
            "['L13-L14']\n",
            "Worst 1%:\n",
            "[]\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 7. Download the predictions\n",
        "\n",
        "#@markdown Once this cell has been executed, a zip-archive with\n",
        "#@markdown the predictions, the corresponding plots and the GMM scores will created and is ready to be downloaded.\n",
        "%%capture\n",
        "final_results = {}\n",
        "\n",
        "predictions = msaSuite.allAlignedPredictions\n",
        "\n",
        "for prot in predictions.keys():\n",
        "  if prot != \"sequence\":\n",
        "      predictions[prot][\"sequence\"] = predictions_single_seq[\"sequence\"][prot]\n",
        "      final_results[prot] = {**predictions[prot]}\n",
        "final_results[\"statistics\"] = jsondata_list[0]\n",
        "\n",
        "json.dump(final_results, open('/content/results/predicted_biophysical_features.json', 'w'), indent=4)\n",
        "\n",
        "!zip -r /content/b2b_tools_results.zip /content/results\n",
        "\n",
        "files.download(\"/content/b2b_tools_results.zip\")"
      ],
      "metadata": {
        "id": "Nckxe9SPemTr",
        "cellView": "form"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Questions? Feedback?\n",
        "\n",
        "Contact us through our [Feedback](https://www.bio2byte.be/b2btools/feedback) page on the bio2Byte website.\n",
        "\n",
        "\n",
        "## PyPi repository\n",
        "\n",
        "Want to try out the b2btools? Download the Bio2Byte's tools package from our PyPi repository: https://pypi.org/project/b2bTools/.\n",
        "\n",
        "## Citations\n",
        "\n",
        "- Implementation of the b2btools to study the protein biophysical features and their conservation:\n",
        "> Kagami, L. P., Orlando, G., Raimondi, D., Ancien, F., Dixit, B., Gavaldá-García, J., Ramasamy, P., Roca-Martínez, J., Tzavella, K., & Vranken, W. (2021). b2bTools: Online predictions for protein biophysical features and their conservation. Nucleic Acids Research, 49(W1), W52–W59. https://doi.org/10.1093/nar/gkab425\n",
        "\n",
        "- DynaMine:\n",
        "> Cilia, E., Pancsa, R., Tompa, P., Lenaerts, T., & Vranken, W. F. (2013). From protein sequence to dynamics and disorder with DynaMine. Nature Communications, 4(1), 2741–2741. https://doi.org/10.1038/ncomms3741\n",
        "\n",
        "- EFoldMine:\n",
        "> Raimondi, D., Orlando, G., Pancsa, R., Khan, T., & Vranken, W. F. (2017). Exploring the Sequence-based Prediction of Folding Initiation Sites in Proteins. Scientific Reports, 7(1), 8826–8826. https://doi.org/10.1038/s41598-017-08366-3\n",
        "\n",
        "- Disomine:\n",
        "> Orlando, G., Raimondi, D., Codicè, F., Tabaro, F., & Vranken, W. (2022). Prediction of Disordered Regions in Proteins with Recurrent Neural Networks and Protein Dynamics. Journal of Molecular Biology, 434(12), 167579. https://doi.org/10.1016/j.jmb.2022.167579\n",
        "\n",
        "- AgMata:\n",
        "> Orlando, G., Silva, A., Macedo-Ribeiro, S., Raimondi, D., & Vranken, W. (2020). Accurate prediction of protein beta-aggregation with generalized statistical potentials. Bioinformatics, 36(7), 2076–2081. https://doi.org/10.1093/bioinformatics/btz912\n",
        "\n",
        "- PSPer:\n",
        "> Orlando, G., Raimondi, D., Tabaro, F., Codicè, F., Moreau, Y., & Vranken, W. F. (2019). Computational identification of prion-like RNA-binding proteins that form liquid phase-separated condensates. Bioinformatics, 35(22), 4617–4623. https://doi.org/10.1093/bioinformatics/btz274"
      ],
      "metadata": {
        "id": "UQiFF_vgqmfq"
      }
    }
  ]
}