{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Bioconda - B2B Tools",
      "provenance": [],
      "collapsed_sections": [],
      "private_outputs": true,
      "machine_shape": "hm",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3.0.4_SingleSequence_demo_bioconda.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Bio2Byte - Single Sequence Predictions\n",
        "For most of below tools, we now have the b2bTools one-stop file upload location where you can easily try them: [online b2BTools](https://bio2byte.be/b2btools/)"
      ],
      "metadata": {
        "id": "EsZPZMEGYV2d"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YaxPffRnrSVW",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Install Conda on this Google Colab notebook \n",
        "#@markdown Running this cell will install the package manager _miniconda_ using the tool [condalab](https://github.com/conda-incubator/condacolab).\n",
        "\n",
        "#@markdown **Notes:** It's important to notice that the running kernel will have been restarted.  \n",
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install_miniconda()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Checking the status of Conda after restarting the kernel \n",
        "#@markdown This cell is just a checkpoint to validate the installation of miniconda.\n",
        "\n",
        "import condacolab\n",
        "condacolab.check()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "kq7sSX0KQWP_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install b2bTools and its dependencies\n",
        "#@markdown After setting up Conda, we are able to proceed with the installation \n",
        "#@markdown of the b2bTools package from the channel ['bioconda'](https://bioconda.github.io/).\n",
        "\n",
        "#@markdown > Bioconda lets you install thousands of software packages related to biomedical research using the conda package manager. \n",
        "\n",
        "#@markdown #### About b2bTools\n",
        "#@markdown In case you want to read further details about this tool, find the documentation and examples on \n",
        "#@markdown our PyPI page: https://pypi.org/project/b2bTools/\n",
        "\n",
        "\n",
        "#@markdown ## Important notes:\n",
        "#@markdown This cell or the following ones might fail after running the previous cells. If that issue happens, run this cell again.\n",
        "#@markdown You could notice this issue by receiving an error message saying:\n",
        "#@markdown > Your session crashed for an unknown reason. View runtime logs\n",
        "!conda install --yes --channel bioconda b2bTools==3.0.4"
      ],
      "metadata": {
        "cellView": "form",
        "id": "ZbFjf8Y6NkAa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Import dependencies\n",
        "#@markdown Run this cell to import the required Python dependencies.\n",
        "\n",
        "import os\n",
        "import json\n",
        "from google.colab import files\n",
        "from b2bTools import SingleSeq\n",
        "from matplotlib import pyplot as plt\n",
        "from matplotlib.lines import Line2D\n",
        "\n",
        "if not os.path.exists(\"/content/results\"):\n",
        "    os.mkdir('/content/results')"
      ],
      "metadata": {
        "id": "cKwsuaFgr3CT",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Input protein sequences\n",
        "%%capture\n",
        "\n",
        "query_sequence_1 = 'YRRRRYYYYYRRRYYYYYYYMESNQSNNGGSGRYYYYYYYNAALNRGGRYVPPHLRGGDGGAAAAASAGGDDRRGGAGGGGYRRGGGNSGGGGGGGYDRGYNDNRDDRDNRG' #@param {type:\"string\"}\n",
        "query_sequence_2 = 'YNNNRGGGGGGYNRQDRGDGGSSNFSRGGYNNRDEGSDNRGSGRDRGYNDNRDDRDNRGGSGGYGRDRNYEDRGYN' #@param {type:\"string\"}\n",
        "query_sequence_3 = 'GRYVPPHLRGGDGGAAAAASAGGDDRRGGAGGGGYRRGGGNSGGGGGGGYDRGYNDNRDDRDNRGGSGGYGRDRNYEDRGYNGGGGGGGNRGY' #@param {type:\"string\"}\n",
        "query_sequence_4 = 'GARDERIEQELFSGQLSGINFDKYEEIPVEATGDDVPQPISLFSDLSLHEGDDRRGGAGGGGYRRGGGNSGGGGGGGYDRGYNDNR' #@param {type:\"string\"}\n",
        "query_sequence_5 = '' #@param {type:\"string\"}\n",
        "query_sequence_6 = '' #@param {type:\"string\"}\n",
        "query_sequence_7 = '' #@param {type:\"string\"}\n",
        "query_sequence_8 = '' #@param {type:\"string\"}\n",
        "query_sequence_9 = '' #@param {type:\"string\"}\n",
        "query_sequence_10 = '' #@param {type:\"string\"}\n",
        "\n",
        "fasta = f'>SEQ_1\\n{query_sequence_1}\\n'\n",
        "\n",
        "if query_sequence_2 != '':\n",
        "    fasta += f'>SEQ_2\\n{query_sequence_2}\\n'\n",
        "if query_sequence_3 != '':\n",
        "    fasta += f'>SEQ_3\\n{query_sequence_3}\\n'\n",
        "if query_sequence_4 != '':\n",
        "    fasta += f'>SEQ_4\\n{query_sequence_4}\\n'\n",
        "if query_sequence_5 != '':\n",
        "    fasta += f'>SEQ_5\\n{query_sequence_5}\\n'\n",
        "if query_sequence_6 != '':\n",
        "    fasta += f'>SEQ_6\\n{query_sequence_6}\\n'\n",
        "if query_sequence_7 != '':\n",
        "    fasta += f'>SEQ_7\\n{query_sequence_7}\\n'\n",
        "if query_sequence_8 != '':\n",
        "    fasta += f'>SEQ_8\\n{query_sequence_8}\\n'\n",
        "if query_sequence_9 != '':\n",
        "    fasta += f'>SEQ_9\\n{query_sequence_9}\\n'\n",
        "if query_sequence_10 != '':\n",
        "    fasta += f'>SEQ_10\\n{query_sequence_10}\\n'\n",
        "\n",
        "with open(\"input_example1.fasta\", \"w\") as f:\n",
        "  f.write(fasta)\n",
        "\n",
        "single_seq = SingleSeq(\"/content/input_example1.fasta\")\n",
        "single_seq.predict(tools=['dynamine', 'efoldmine', 'disomine', 'agmata'])\n",
        "\n",
        "all_predictions = single_seq.get_all_predictions()\n",
        "keys = list(all_predictions.keys())\n",
        "\n",
        "max_seq_len = max([len(pred['seq']) for pred in all_predictions.values()])"
      ],
      "metadata": {
        "id": "a4AYPR-ByvL4",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Plot Single Sequence predictions\n",
        "#@markdown The predictions reflect 'emerging' properties, so what the sequence \n",
        "#@markdown is capable of, not necessarily what it will do in a particular \n",
        "#@markdown context, for example when it adopts a specific fold.\n",
        "\n",
        "#@markdown #### DynaMine backbone dynamics\n",
        "#@markdown > Values above 0.8 indicate rigid conformations, \n",
        "#@markdown > values above 1.0 membrane spanning regions, \n",
        "#@markdown > values below 0.69 flexible regions. \n",
        "#@markdown > Values between 0.69-0.80 are 'context' dependent and capable of\n",
        "#@markdown > being either rigid or flexible.\n",
        "\n",
        "#@markdown #### DynaMine sidechain dynamics\n",
        "#@markdown > Higher values mean more likely rigid. \n",
        "#@markdown > These values are highly dependent on the amino acid type \n",
        "#@markdown > (i.e. a Trp will be rigid, an Asp flexible).\n",
        "\n",
        "#@markdown #### DynaMine conformational propensities (sheet, helix, coil, ppII (polyproline II))\n",
        "#@markdown > Higher values indicate higher propensities.\n",
        "\n",
        "#@markdown #### EFoldMine earlyFolding propensity\n",
        "#@markdown > Values above 0.169 indicate residues that are likely to start \n",
        "#@markdown > the protein folding process, based on only local interactions \n",
        "#@markdown > with other amino acids.\n",
        "\n",
        "#@markdown #### Disomine disorder\n",
        "#@markdown >  Values above 0.5 indicate that this is likely a disordered residue.\n",
        "\n",
        "\n",
        "fig, axs = plt.subplots(2, 4)\n",
        "\n",
        "ax1 = axs[0, 0]\n",
        "ax2 = axs[0, 1]\n",
        "ax3 = axs[0, 2]\n",
        "ax4 = axs[0, 3]\n",
        "ax5 = axs[1, 0]\n",
        "ax6 = axs[1, 1]\n",
        "ax7 = axs[1, 2]\n",
        "ax8 = axs[1, 3]\n",
        "\n",
        "fig.set_figwidth(30)\n",
        "fig.set_figheight(10)\n",
        "\n",
        "fig.suptitle(f'Single Sequence Predictions')\n",
        "\n",
        "for seq_key in keys:\n",
        "    predictions = all_predictions[seq_key]\n",
        "    x_position = range(len(predictions['seq']))\n",
        "\n",
        "    backbone_pred = predictions['backbone']\n",
        "    coil_pred = predictions['coil']\n",
        "    sheet_pred = predictions['sheet']\n",
        "    ppII_pred = predictions['ppII']\n",
        "    helix_pred = predictions['helix']\n",
        "    sidechain_pred = predictions['sidechain']\n",
        "    disomine_pred = predictions['disoMine']\n",
        "    earlyFolding_pred = predictions['earlyFolding'] \n",
        "    agmata_pred = predictions['agmata']\n",
        "\n",
        "    ax1.plot(x_position, backbone_pred, label=seq_key)\n",
        "    ax2.plot(x_position, sidechain_pred, label=seq_key)\n",
        "    ax3.plot(x_position, coil_pred, label=seq_key)\n",
        "    ax4.plot(x_position, sheet_pred, label=seq_key)\n",
        "    ax5.plot(x_position, ppII_pred, label=seq_key)\n",
        "    ax6.plot(x_position, helix_pred, label=seq_key)\n",
        "    ax7.plot(x_position, disomine_pred, label=seq_key)\n",
        "    ax8.plot(x_position, earlyFolding_pred, label=seq_key)\n",
        "\n",
        "ax1.set_title('DynaMine backbone dynamics')\n",
        "ax1.set_ylim([-0.2, 1.2])\n",
        "ax1.set_xlabel('residue index')\n",
        "ax1.set_ylabel('prediction values')\n",
        "ax1.axhspan(1, 1.2, alpha=0.3, color='red')\n",
        "ax1.axhspan(0.8, 1, alpha=0.5, color='pink')\n",
        "ax1.axhspan(0.69, 0.8, alpha=0.5, color='orange')\n",
        "ax1.axhspan(-0.2, 0.69, alpha=0.5, color='yellow')\n",
        "ax1.grid(axis='y')\n",
        "ax1.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "\n",
        "ax2.set_title('DynaMine sidechain dynamics')\n",
        "ax2.set_ylim([-0.2, 1.2])\n",
        "ax2.set_xlabel('residue index')\n",
        "ax2.set_ylabel('prediction values')\n",
        "ax2.grid(axis='y')\n",
        "ax2.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "\n",
        "ax3.set_title('DynaMine conformational propensities: Coil')\n",
        "ax3.set_ylim([-0.2, 1.2])\n",
        "ax3.set_xlabel('residue index')\n",
        "ax3.set_ylabel('prediction values')\n",
        "ax3.grid(axis='y')\n",
        "ax3.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "\n",
        "ax4.set_title('DynaMine conformational propensities: Sheet')\n",
        "ax4.set_ylim([-0.2, 1.2])\n",
        "ax4.set_xlabel('residue index')\n",
        "ax4.set_ylabel('prediction values')\n",
        "ax4.grid(axis='y')\n",
        "ax4.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "ax5.set_title('DynaMine conformational propensities: ppII (polyproline II)')\n",
        "ax5.set_ylim([-0.2, 1.2])\n",
        "ax5.set_xlabel('residue index')\n",
        "ax5.set_ylabel('prediction values')\n",
        "ax5.grid(axis='y')\n",
        "ax5.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "\n",
        "ax6.set_title('DynaMine conformational propensities: Helix')\n",
        "ax6.set_ylim([-0.2, 1.2])\n",
        "ax6.set_xlabel('residue index')\n",
        "ax6.set_ylabel('prediction values')\n",
        "ax6.grid(axis='y')\n",
        "ax6.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "\n",
        "ax7.set_title('Early folding (EFoldMine)')\n",
        "ax7.set_ylim([-0.2, 1.2])\n",
        "ax7.set_xlabel('residue index')\n",
        "ax7.set_ylabel('prediction values')\n",
        "ax7.axhspan(-0.2, 0.169, alpha=0.5, color='yellow')\n",
        "ax7.axhspan(0.169, 1.2, alpha=0.5, color='orange')\n",
        "ax7.grid(axis='y')\n",
        "ax7.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "\n",
        "ax8.set_title('Disorder (disoMine)')\n",
        "ax8.set_ylim([-0.2, 1.2])\n",
        "ax8.set_xlabel('residue index')\n",
        "ax8.set_ylabel('prediction values')\n",
        "ax8.axhspan(0.5, 1.2, alpha=0.5, color='orange')\n",
        "ax8.axhspan(-0.2, 0.5, alpha=0.5, color='yellow')\n",
        "ax8.grid(axis='y')\n",
        "ax8.set_xlim([0, max_seq_len - 1])\n",
        "\n",
        "plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.3), fancybox=True, shadow=True, ncol=len(keys))\n",
        "plt.subplots_adjust(hspace=0.4)\n",
        "plt.savefig('/content/results/single_sequence_predictons.png')\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "KOH6FzXFrzw4",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Plot AgMata aggregation propensity\n",
        "#@markdown The predictions reflect 'emerging' properties, so what the sequence \n",
        "#@markdown is capable of, not necessarily what it will do in a particular \n",
        "#@markdown context, for example when it adopts a specific fold.\n",
        "\n",
        "#@markdown > These values are divided by a factor of 20 from the original. \n",
        "#@markdown > Peaks indicate residues likely to be involved in beta-sheet \n",
        "#@markdown > aggregation.\n",
        "\n",
        "fig, ax = plt.subplots(1, 1)\n",
        "fig.set_figwidth(30)\n",
        "fig.set_figheight(5)\n",
        "fig.suptitle('Agmata aggregation propensity')\n",
        "\n",
        "for seq_key in keys:\n",
        "    predictions = all_predictions[seq_key]\n",
        "    agmata_pred = predictions['agmata']\n",
        "    ax.plot(range(len(agmata_pred)), agmata_pred, label=seq_key)\n",
        "\n",
        "ax.set_xlim([0, max_seq_len - 1])\n",
        "ax.set_xlabel('residue index')\n",
        "ax.set_ylabel('prediction values')\n",
        "ax.grid(axis='y')\n",
        "\n",
        "plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.2), fancybox=True, shadow=True, ncol=len(keys))\n",
        "plt.savefig('/content/results/agmata_predictons.png')\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "m_K9OQKnSXg3",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Plot Single Sequence predictions (3D projection)\n",
        "#@markdown The predictions reflect 'emerging' properties, so what the sequence \n",
        "#@markdown is capable of, not necessarily what it will do in a particular \n",
        "#@markdown context, for example when it adopts a specific fold.\n",
        "\n",
        "#@markdown #### DynaMine backbone dynamics\n",
        "#@markdown > Values above 0.8 indicate rigid conformations, \n",
        "#@markdown > values above 1.0 membrane spanning regions, \n",
        "#@markdown > values below 0.69 flexible regions. \n",
        "#@markdown > Values between 0.69-0.80 are 'context' dependent and capable of\n",
        "#@markdown > being either rigid or flexible.\n",
        "\n",
        "#@markdown #### DynaMine sidechain dynamics\n",
        "#@markdown > Higher values mean more likely rigid. \n",
        "#@markdown > These values are highly dependent on the amino acid type \n",
        "#@markdown > (i.e. a Trp will be rigid, an Asp flexible).\n",
        "\n",
        "#@markdown #### DynaMine conformational propensities (sheet, helix, coil, ppII (polyproline II))\n",
        "#@markdown > Higher values indicate higher propensities.\n",
        "\n",
        "#@markdown #### EFoldMine earlyFolding propensity\n",
        "#@markdown > Values above 0.169 indicate residues that are likely to start \n",
        "#@markdown > the protein folding process, based on only local interactions \n",
        "#@markdown > with other amino acids.\n",
        "\n",
        "#@markdown #### Disomine disorder\n",
        "#@markdown >  Values above 0.5 indicate that this is likely a disordered residue.\n",
        "\n",
        "fig = plt.figure(figsize=plt.figaspect(0.5))\n",
        "# axs = fig.add_subplot(2, 4, 1, projection='3d')\n",
        "\n",
        "ax1 = fig.add_subplot(4, 2, 1, projection='3d')\n",
        "ax2 = fig.add_subplot(4, 2, 2, projection='3d')\n",
        "ax3 = fig.add_subplot(4, 2, 3, projection='3d')\n",
        "ax4 = fig.add_subplot(4, 2, 4, projection='3d')\n",
        "ax5 = fig.add_subplot(4, 2, 5, projection='3d')\n",
        "ax6 = fig.add_subplot(4, 2, 6, projection='3d')\n",
        "ax7 = fig.add_subplot(4, 2, 7, projection='3d')\n",
        "ax8 = fig.add_subplot(4, 2, 8, projection='3d')\n",
        "\n",
        "fig.set_figwidth(30)\n",
        "fig.set_figheight(40)\n",
        "\n",
        "fig.suptitle(f'Single Sequence Predictions (3D Projection)')\n",
        "\n",
        "for index, seq_key in enumerate(keys):\n",
        "    predictions = all_predictions[seq_key]\n",
        "    x_position = range(len(predictions['seq']))\n",
        "    z_position = [index] * len(predictions['seq'])\n",
        "\n",
        "    backbone_pred = predictions['backbone']\n",
        "    coil_pred = predictions['coil']\n",
        "    sheet_pred = predictions['sheet']\n",
        "    ppII_pred = predictions['ppII']\n",
        "    helix_pred = predictions['helix']\n",
        "    sidechain_pred = predictions['sidechain']\n",
        "    disomine_pred = predictions['disoMine']\n",
        "    earlyFolding_pred = predictions['earlyFolding'] \n",
        "    agmata_pred = predictions['agmata']\n",
        "\n",
        "    ax1.plot(x_position, z_position, backbone_pred, label=seq_key)\n",
        "    ax2.plot(x_position, z_position, sidechain_pred, label=seq_key)\n",
        "    ax3.plot(x_position, z_position, coil_pred, label=seq_key)\n",
        "    ax4.plot(x_position, z_position, sheet_pred, label=seq_key)\n",
        "    ax5.plot(x_position, z_position, ppII_pred, label=seq_key)\n",
        "    ax6.plot(x_position, z_position, helix_pred, label=seq_key)\n",
        "    ax7.plot(x_position, z_position, disomine_pred, label=seq_key)\n",
        "    ax8.plot(x_position, z_position, earlyFolding_pred, label=seq_key)\n",
        "\n",
        "ax1.set_title('DynaMine backbone dynamics')\n",
        "ax1.set_zlim([-0.1, 1.1])\n",
        "ax1.set_xlabel('residue index')\n",
        "ax1.set_zlabel('prediction values')\n",
        "ax1.set_yticks(range(len(keys)))\n",
        "ax1.set_yticklabels(keys)\n",
        "ax1.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax2.set_title('DynaMine sidechain dynamics')\n",
        "ax2.set_zlim([-0.1, 1.1])\n",
        "ax2.set_xlabel('residue index')\n",
        "ax2.set_zlabel('prediction values')\n",
        "ax2.set_yticks(range(len(keys)))\n",
        "ax2.set_yticklabels(keys)\n",
        "ax2.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax3.set_title('DynaMine conformational propensities: Coil')\n",
        "ax3.set_zlim([-0.1, 1.1])\n",
        "ax3.set_xlabel('residue index')\n",
        "ax3.set_zlabel('prediction values')\n",
        "ax3.set_yticks(range(len(keys)))\n",
        "ax3.set_yticklabels(keys)\n",
        "ax3.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax4.set_title('DynaMine conformational propensities: Sheet')\n",
        "ax4.set_zlim([-0.1, 1.1])\n",
        "ax4.set_xlabel('residue index')\n",
        "ax4.set_zlabel('prediction values')\n",
        "ax4.set_yticks(range(len(keys)))\n",
        "ax4.set_yticklabels(keys)\n",
        "ax4.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax5.set_title('DynaMine conformational propensities: ppII (polyproline II)')\n",
        "ax5.set_zlim([-0.1, 1.1])\n",
        "ax5.set_xlabel('residue index')\n",
        "ax5.set_zlabel('prediction values')\n",
        "ax5.set_yticks(range(len(keys)))\n",
        "ax5.set_yticklabels(keys)\n",
        "ax5.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax6.set_title('DynaMine conformational propensities: Helix')\n",
        "ax6.set_zlim([-0.1, 1.1])\n",
        "ax6.set_xlabel('residue index')\n",
        "ax6.set_ylabel('sequence')\n",
        "ax6.set_zlabel('prediction values')\n",
        "ax6.set_yticks(range(len(keys)))\n",
        "ax6.set_yticklabels(keys)\n",
        "ax6.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax7.set_title('Disorder (disoMine)')\n",
        "ax7.set_zlim([-0.1, 1.1])\n",
        "ax7.set_xlabel('residue index')\n",
        "ax7.set_ylabel('sequence')\n",
        "ax7.set_zlabel('prediction values')\n",
        "ax7.set_yticks(range(len(keys)))\n",
        "ax7.set_yticklabels(keys)\n",
        "ax7.view_init(60, -45)\n",
        "\n",
        "\n",
        "ax8.set_title('Early folding (EFoldMine)')\n",
        "ax8.set_zlim([-0.1, 1.1])\n",
        "ax8.set_xlabel('residue index')\n",
        "ax8.set_ylabel('sequence')\n",
        "ax8.set_zlabel('prediction values')\n",
        "ax8.set_yticks(range(len(keys)))\n",
        "ax8.set_yticklabels(keys)\n",
        "ax8.view_init(60, -45)\n",
        "\n",
        "\n",
        "plt.subplots_adjust(hspace=0.2)\n",
        "plt.savefig('/content/results/single_sequence_predictons_3d.png')\n",
        "plt.show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "JKVklmzvY9Y3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Plot AgMata aggregation propensity (3D projection)\n",
        "#@markdown The predictions reflect 'emerging' properties, so what the sequence \n",
        "#@markdown is capable of, not necessarily what it will do in a particular \n",
        "#@markdown context, for example when it adopts a specific fold.\n",
        "\n",
        "#@markdown > These values are divided by a factor of 20 from the original. \n",
        "#@markdown > Peaks indicate residues likely to be involved in beta-sheet \n",
        "#@markdown > aggregation.\n",
        "\n",
        "fig = plt.figure(figsize=plt.figaspect(0.5))\n",
        "ax1 = fig.add_subplot(1, 1, 1, projection='3d')\n",
        "\n",
        "fig.set_figwidth(15)\n",
        "fig.set_figheight(10)\n",
        "\n",
        "fig.suptitle(f'Single Sequence Predictions (3D Projection)')\n",
        "\n",
        "for index, seq_key in enumerate(keys):\n",
        "    predictions = all_predictions[seq_key]\n",
        "    x_position = range(len(predictions['seq']))\n",
        "    z_position = [index] * len(predictions['seq'])\n",
        "\n",
        "    agmata_pred = predictions['agmata']\n",
        "\n",
        "    ax1.plot(x_position, z_position, agmata_pred, label=seq_key)\n",
        "\n",
        "ax1.set_title('Agmata aggregation propensity')\n",
        "ax1.set_xlabel('residue index')\n",
        "ax1.set_zlabel('prediction values')\n",
        "ax1.set_yticks(range(len(keys)))\n",
        "ax1.set_yticklabels(keys)\n",
        "ax1.view_init(60, -45)\n",
        "\n",
        "plt.savefig('/content/results/agmata_predictons_3d.png')\n",
        "plt.show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "oRxE84upgFdn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download the results\n",
        "#@markdown After running this cell you will be able to download the plots\n",
        "#@markdown plain results in JSON format to your computer.\n",
        "\n",
        "json.dump(all_predictions, open('/content/results/b2btools_results.json', 'w'), indent=4, sort_keys=True)\n",
        "\n",
        "!zip -r /content/b2b-tools-results.zip /content/results\n",
        "\n",
        "files.download(\"/content/b2b-tools-results.zip\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "NyBFvJ4vMqBJ"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}
