# Public Notebooks
Welcome to the repository of Jupyter Notebooks created by the Bio2Byte group!

<p align="center">
  <img src="https://pbs.twimg.com/profile_images/1247824923546079232/B9b_Yg7n_400x400.jpg" width="48px"/>
  <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/31/Vrije_Universiteit_Brussel_logo.svg/1200px-Vrije_Universiteit_Brussel_logo.svg.png" alt="vub" width="120px"/>
</p>

## About Bio2Byte

Proteins are the molecular machines that make cells work. They perform a wide variety of functions through interactions with each other and many additional molecules. Traditionally, proteins are described in a single static state (a picture). It is now increasingly recognised that many proteins can adopt multiple states and move between these conformational states dynamically (a movie).

We investigate how the dynamics, conformational states and available experimental data of proteins relates to their amino acid sequence. Underlying physical and chemical principles are computationally unravelled through data integration, analysis and machine learning, so connecting them to biological events and improving our understanding of the way proteins work.

The Bio2Byte group is primarily situated at the Interuniversity Institute of Bioinformatics in Brussels, a collaborative interfaculty institute between the Vrije Universiteit Brussel and the Université Libre de Bruxelles. It is located at the the ULB side of the Pleinlaan/La Plaine campus on the 6th floor of the C building.

At the VUB, the group is linked to Structural Biology Brussels at the Bioengineering sciences department, as well as the departments of Computer Science and Chemistry at the Faculty of Sciences, plus to the Biomedical sciences at the Faculty of Medicine and Pharmacy.

### 🔗 Links
- [Bio2Byte](https://bio2byte.be)
- [Bio2Byte online predictors](https://bio2byte.be/b2btools)
- [Feedaback](https://www.bio2byte.be/b2btools/feedback)
- [Bio2Byte tools package](https://pypi.org/project/b2bTools/)

## 📚 List of available notebooks

### Single Sequence analysis

Explore prediction-based 'biophysical variation' of proteins from up to 10 sequences at the same time.

#### 📦 Using PyPI (Python Package Index)

##### 📓 Title: `Bio2ByteTools_v3_singleseq_pypi.ipynb` (latest version)

Example of usage of our Bio2Byte tools package for single sequence file inputs (in FASTA format) using v3.0.5

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3.0.5_SingleSequence_demo.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3.0.5_SingleSequence_demo.ipynb)

#### Title: `Bio2ByteTools_v3_singleseq_pypi_with_files.ipynb` (latest version supporting FASTA files as input)

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3_singleseq_pypi_with_files.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3_singleseq_pypi_with_files.ipynb)

##### 📓 Title: `Bio2ByteTools_v3_singleseq_pypi.ipynb` (version including PSP predictions)

Example of usage of our Bio2Byte tools package for single sequence file inputs (in FASTA format) using v3.0.5

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3_singleseq_pypi.ipynbb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3_singleseq_pypi.ipynb)

##### 📓 Title: `AlphaFold_+_B2B_Tools.ipynb`

Run a set of biophysical feature predictions on top of an Alphafold thee-dimensional predicted structure.

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/AlphaFold_%2B_B2B_Tools.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/AlphaFold_%2B_B2B_Tools.ipynb)

##### 📓 Title: `Bio2Byte_AlphaFold_Analysis.ipynb`

From the output files generated from Alphafold, this notebook runs a set of biophysical feature prediction

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2Byte_AlphaFold_Analysis.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2Byte_AlphaFold_Analysis.ipynb)

#### 📦 Using Conda package manager

##### 📓 Title: `Bio2ByteTools_v3.0.5_SingleSequence_demo_bioconda.ipynb` (latest version)

Example of usage of our Bio2Byte tools package for single sequence file inputs (in FASTA format) using v3.0.5 from Bioconda channel

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3.0.5_SingleSequence_demo_bioconda.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3.0.5_SingleSequence_demo_bioconda.ipynb)

### Multiple Sequence Alignment analysis

Explore prediction-based 'biophysical variation' of proteins from a multiple sequence alignment.

#### 📦 Using PyPI (Python Package Index)

##### 📓 Title: `Bio2ByteTools_v3_multipleseq_pypi.ipynb`

Example of usage of our Bio2Byte tools package for Multiple Sequence Alignment file inputs

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3_multipleseq_pypi.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2ByteTools_v3_multipleseq_pypi.ipynb)

### PyMol on Google Colab

##### 📓 Title: `Bio2Byte_Pymol_scripting.ipynb`

Example of PyMol scripting in Python. This notebook fetchs a PDB and renders both a static image and a 360-degree rotated animation

- [📖 Read Jupyter Notebook](https://github.com/Bio2Byte/public_notebooks/blob/main/Bio2Byte_Pymol_scripting.ipynb)
- [⏯️ Open Jupyter Notebook on Google Colab](https://colab.research.google.com/github/Bio2Byte/public_notebooks/blob/main/Bio2Byte_Pymol_scripting.ipynb)


## 📃 How to use them?
1. Open the Jupyter notebook file of interest
1. There will be a "Open In Colab" button, if you click on it, a new session on Google Colab will be started
1. From the Google Colab session you will be able to run the code of the notebook.

## Copyright

© Wim Vranken, Bio2Byte group, VUB
